# Prep SNP + Indel Call



```bash

# Get all bam files from Paleomix run
# Call SNPs with GATK4 - Haplotype Caller

# split it up into pop batches , but run invidivually

for f in $(cat Hybrid.samples | grep "SPI" | tail -n+2 ) ; 
do echo ${f};
for i in {01..25} Z Unplaced ;
do echo ${i} ;
sbatch nucl_SNPs_pt1a.sh ${f} ${i} ; 
done ;
echo "W" ;
sbatch nucl_SNPs_pt1b.sh ${f} ;
done

for f in $(cat Hybrid.samples | grep "BJO" ) ; 
do echo ${f};
for i in {01..25} Z Unplaced ;
do echo ${i} ;
sbatch nucl_SNPs_pt1a.sh ${f} ${i} ; 
done ;
echo "W" ;
sbatch nucl_SNPs_pt1b.sh ${f} ;
done

for f in $(cat Hybrid.samples | grep "ROS" ) ; 
do echo ${f};
for i in {01..25} Z Unplaced ;
do echo ${i} ;
sbatch nucl_SNPs_pt1a.sh ${f} ${i} ; 
done ;
echo "W" ;
sbatch nucl_SNPs_pt1b.sh ${f} ;
done

```

nucl_SNPs_pt1a.sh

```bash



module load GATK/4.2.0.0-GCCcore-9.3.0-Java-11

gatk --java-options "-Xmx8g" HaplotypeCaller \
-VS STRICT \
-R Puffin_ReferenceGenome.NU.MT.fasta \
-I ${1}.Puffin_ReferenceGenome.NU.MT.realigned.bam \
-ERC GVCF \
-L Chr_${2} \
-O ${1}.Chr_${2}.g.vcf.gz 2> Haplotype_caller.${1}.Chr_${2}.out

mv Haplotype_caller.${1}.Chr_${2}.out Log/ ;
mv ${1}.Chr_${2}.g.vcf.gz Data/ ;
mv ${1}.Chr_${2}.g.vcf.gz.tbi Data/ ;

```

nanoÂ nucl_SNPs_pt1b.sh

```bash

module load GATK/4.2.0.0-GCCcore-9.3.0-Java-11

gatk --java-options "-Xmx8g" HaplotypeCaller \
-VS STRICT \
-R Puffin_ReferenceGenome.NU.MT.fasta \
-I ${1}.Puffin_ReferenceGenome.NU.MT.realigned.bam \
-ploidy 1 \
-ERC GVCF \
-L Chr_W \
-O ${1}.Chr_W.g.vcf.gz 2> Haplotype_caller.${1}.Chr_W.out

mv Haplotype_caller.${1}.Chr_W.out Log/ ;
mv ${1}.Chr_W.g.vcf.gz Data/ ;
mv ${1}.Chr_W.g.vcf.gz.tbi Data/ ;

```
