# SNP Call

```bash

for i in {01..25} Z Unplaced W ;
do echo ${i} ;
touch SPI_BJO_ROS.puffin.sample_map.Chr_${i} ;
done

cat Hybrid.samples | while read line ;
do echo ${line} ;
for i in {01..25} Z Unplaced W ;
do echo -e "${line}\t${line}.Chr_${i}.g.vcf.gz" >> SPI_BJO_ROS.puffin.sample_map.Chr_${i} ;
done ;
done ;

cat SPI_BJO_ROS.puffin.sample_map.Chr_W | cut -f2 > Chr_W.gvcf.list

for i in {01..25} Z Unplaced ; 
do echo ${i} ;
sbatch nucl_SNPs_pt2a.sh ${i} ; 
done
sbatch nucl_SNPs_pt2b.sh
```

Combine GVCFs and Call Genotypes (GATK4) - per chromosome

nucl_SNPs_pt2a.sh

```bash

module load GATK/4.2.0.0-GCCcore-9.3.0-Java-11

mkdir tmp_${1}

gatk --java-options "-Xmx8g -Xms8g" GenomicsDBImport \
-L Chr_${1} \
--genomicsdb-workspace-path SPI_BJO_ROS.Puffin_NU_GVCF_Chr_${1}.database \
--sample-name-map SPI_BJO_ROS.puffin.sample_map.Chr_${1} \
--tmp-dir tmp_${1}

gatk --java-options "-Xmx8g" GenotypeGVCFs \
-R Puffin_ReferenceGenome.NU.MT.fasta \
-L Chr_${1} \
-V gendb://SPI_BJO_ROS.Puffin_NU_GVCF_Chr_${1}.database \
-O SPI_BJO_ROS.Chr_${1}.raw.vcf.gz \
--tmp-dir tmp_${1}

rm -r tmp_${1}

```

nucl_SNPs_pt2b.sh

```bash

module load GATK/4.2.0.0-GCCcore-9.3.0-Java-11

mkdir tmp_Chr_W

gatk --java-options "-Xmx8g" CombineGVCFs \
-R Puffin_ReferenceGenome.NU.MT.fasta \
-L Chr_W \
-V Chr_W.gvcf.list \
-O SPI_BJO_ROS.Chr_W.All.g.vcf.gz \
--tmp-dir tmp_Chr_W

gatk --java-options "-Xmx8g" GenotypeGVCFs \
-R Puffin_ReferenceGenome.NU.MT.fasta \
-L Chr_W \
-V SPI_BJO_ROS.Chr_W.All.g.vcf.gz \
-O SPI_BJO_ROS.Chr_W.raw.vcf.gz \
--tmp-dir tmp_Chr_W

rm -r tmp_Chr_W

```

Combine VCF’s of various chromosomes

```bash

module load BCFtools/1.9-intel-2018b

ls | grep "raw.vcf.gz" | grep -v "Chr_W" | grep -v "Chr_Z" | grep -v "Chr_Unplaced" | grep -v ".tbi" > vcf.list
ls | grep "raw.vcf.gz" | grep -v "Chr_W" | grep -v ".tbi" > vcf_All.list

cat vcf.list | while read line ;
do echo ${line} ;
bcftools index ${line} ;
done ;
bcftools index SPI_BJO_ROS.Chr_W.raw.vcf.gz
bcftools index SPI_BJO_ROS.Chr_Z.raw.vcf.gz
bcftools index SPI_BJO_ROS.Chr_Unplaced.raw.vcf.gz

bcftools concat -n -O z -o SPI_BJO_ROS_NU_SnpsAndIndels.Autosomes.raw.vcf.gz -f vcf.list
#-n naive-> faster
#-O output format gzipped vcf
bcftools index SPI_BJO_ROS_NU_SnpsAndIndels.Autosomes.raw.vcf.gz

bcftools concat -n -O z -o SPI_BJO_ROS_NU_SnpsAndIndels.AllNoW.raw.vcf.gz -f vcf_All.list
bcftools index SPI_BJO_ROS_NU_SnpsAndIndels.AllNoW.raw.vcf.gz

```

NON VARIANT SITES

Generate VCF with non-variant sites

```bash

for i in {01..24} Z Unplaced ; 
do echo ${i} ;
sbatch nucl_SNPs_pt2a_edit.sh ${i} ; 
done
sbatch nucl_SNPs_pt2b_edit.sh

```

Combine GVCFs and Call Genotypes (GATK4) - per chromosome

nucl_SNPs_pt2a_edit.sh

```bash


module load GATK/4.2.0.0-GCCcore-9.3.0-Java-11

mkdir tmp_${1}

gatk --java-options "-Xmx8g" GenotypeGVCFs \
-R Puffin_ReferenceGenome.NU.MT.fasta \
-L Chr_${1} \
-V gendb://SPI_BJO_ROS.Puffin_NU_GVCF_Chr_${1}.database \
-O SPI_BJO_ROS.Chr_${1}.raw.withnonvariant.vcf.gz \
--include-non-variant-sites \
--tmp-dir tmp_${1}


```

nucl_SNPs_pt2b_edit.sh

```bash


module load GATK/4.2.0.0-GCCcore-9.3.0-Java-11

mkdir tmp_Chr_W

gatk --java-options "-Xmx8g" GenotypeGVCFs \
-R Puffin_ReferenceGenome.NU.MT.fasta \
-L Chr_W \
-V SPI_BJO_ROS.Chr_W.All.g.vcf.gz \
-O SPI_BJO_ROS.Chr_W.raw.withnonvariant.vcf.gz \
--include-non-variant-sites \
--tmp-dir tmp_Chr_W


```

Combine VCF’s of various chromosomes

```bash

module load BCFtools/1.9-intel-2018b

ls | grep "raw.withnonvariant.vcf.gz" | grep -v "Chr_W" | grep -v "Chr_Z" | grep -v "Chr_Unplaced" | grep -v ".tbi" > vcf.list
ls | grep "raw.withnonvariant.vcf.gz" | grep -v "Chr_W" | grep -v ".tbi" > vcf_All.list

cat vcf.list | while read line ;
do echo ${line} ;
bcftools index ${line} ;
done ;
bcftools index SPI_BJO_ROS.Chr_W.raw.withnonvariant.vcf.gz
bcftools index SPI_BJO_ROS.Chr_Z.raw.withnonvariant.vcf.gz
bcftools index SPI_BJO_ROS.Chr_Unplaced.raw.withnonvariant.vcf.gz

bcftools concat -n -O z -o SPI_BJO_ROS_NU_SnpsAndIndels.Autosomes.raw.withnonvariant.vcf.gz -f vcf.list
#-n naive-> faster
#-O output format gzipped vcf
bcftools index SPI_BJO_ROS_NU_SnpsAndIndels.Autosomes.raw.withnonvariant.vcf.gz

bcftools concat -n -O z -o SPI_BJO_ROS_NU_SnpsAndIndels.AllNoW.raw.withnonvariant.vcf.gz -f vcf_All.list
bcftools index SPI_BJO_ROS_NU_SnpsAndIndels.AllNoW.raw.withnonvariant.vcf.gz

```