# GATK filter

Filter of VCF 

```bash

#Total Variants
zcat SPI_BJO_ROS_NU_SnpsAndIndels.AllNoW.raw.vcf.gz | grep -v "#" | wc -l

#Only SNPs 
module purge
module load VCFtools/0.1.16-intel-2018b-Perl-5.28.0
module load BCFtools/1.9-intel-2018b
vcftools --gzvcf SPI_BJO_ROS_NU_SnpsAndIndels.AllNoW.raw.vcf.gz --remove-indels --out SPI_BJO_ROS_NU_Snps.AllNoW.raw



#Apply GATK filters
bcftools filter --SnpGap 10 -e 'QD < 2.0 || MQ < 40 || FS > 60.0 || SOR > 3 || MQRankSum < -12.5 || MQRankSum > 12.5 || ReadPosRankSum > 8.0 || ReadPosRankSum < -8.0' -O z -o SPI_BJO_ROS_NU_SnpsAndIndels.AllNoW.filtered1.vcf.gz SPI_BJO_ROS_NU_SnpsAndIndels.AllNoW.raw.vcf.gz
bcftools view -H SPI_BJO_ROS_NU_SnpsAndIndels.AllNoW.filtered1.vcf.gz | wc -l 
bcftools view -v snps -H SPI_BJO_ROS_NU_SnpsAndIndels.AllNoW.filtered1.vcf.gz | wc -l

#the following step is kind of unneccessary because we are filtering for DP and GQ later
bcftools filter -S . -e 'FMT/DP<3 | FMT/GQ<15' -O z -o SPI_BJO_ROS_NU_SnpsAndIndels.AllNoW.filtered2.vcf.gz SPI_BJO_ROS_NU_SnpsAndIndels.AllNoW.filtered1.vcf.gz

ml purge
ml Miniconda3/4.9.2
conda activate miniconda3_envs/vcflib
gunzip SPI_BJO_ROS_NU_SnpsAndIndels.AllNoW.filtered2.vcf.gz
vcffixup SPI_BJO_ROS_NU_SnpsAndIndels.AllNoW.filtered2.vcf | gzip - > SPI_BJO_ROS_NU_SnpsAndIndels.AllNoW.filtered2.edit.vcf.gz
gzip SPI_BJO_ROS_NU_SnpsAndIndels.AllNoW.filtered2.vcf
conda deactivate

bcftools view -H SPI_BJO_ROS_NU_SnpsAndIndels.AllNoW.filtered2.edit.vcf.gz | wc -l 
bcftools view -v snps -H SPI_BJO_ROS_NU_SnpsAndIndels.AllNoW.filtered2.edit.vcf.gz | wc -l

# Exclude monomorphic SNPs and Indels and only include biallelic
vcftools --gzvcf SPI_BJO_ROS_NU_SnpsAndIndels.AllNoW.filtered2.edit.vcf.gz --remove-indels --min-alleles 2 --max-alleles 2 --maf 0.05 --max-maf 0.95 --recode --recode-INFO-all --stdout | gzip > SPI_BJO_ROS_NU_Snps.AllNoW.vcf.gz
zcat SPI_BJO_ROS_NU_Snps.AllNoW.vcf.gz | grep -v "#" | wc -l

vcftools --gzvcf SPI_BJO_ROS_NU_Snps.AllNoW.vcf.gz --not-chr Chr_Z --recode --recode-INFO-all --stdout | gzip > SPI_BJO_ROS_NU_Snps.Autosomes.vcf.gz

mv out.log NoZ.out.log

```


Puffin VCF: Filter on Depth, Missingness, HWE etc.

```bash


module purge
module load VCFtools/0.1.16-intel-2018b-Perl-5.28.0
module load BCFtools/1.9-intel-2018b

#GLOBAL DEPTH
#maxDP 500
#minDP 250

bcftools filter -e 'INFO/DP[*] > 500 | INFO/DP[*] < 250 ' -O z -o SPI_BJO_ROS_NU_Snps.Autosomes.filtered1.vcf.gz SPI_BJO_ROS_NU_Snps.Autosomes.vcf.gz
zcat SPI_BJO_ROS_NU_Snps.Autosomes.filtered1.vcf.gz | grep -v "#" | wc -l

#OTHER FILTERS
#--max-missing 1
#--max-meanDP 43.2 
#--min-meanDP 10.8
#--minQ 30

vcftools --gzvcf SPI_BJO_ROS_NU_Snps.Autosomes.filtered1.vcf.gz --max-missing 1 --max-meanDP 43.2 --min-meanDP 10.8 --minQ 30 --remove-filtered-all --recode --recode-INFO-all --stdout | gzip > SPI_BJO_ROS_NU_Snps.Autosomes.filtered2.vcf.gz

mv out.log Autosomes.filtered2.log

vcftools --gzvcf SPI_BJO_ROS_NU_Snps.Autosomes.filtered2.vcf.gz --not-chr Chr_Unplaced --recode --recode-INFO-all --stdout | gzip > SPI_BJO_ROS_NU_Snps.Autosomes.filtered5.vcf.gz
mv out.log NoZ_NoUnpl_noHWE.out.log
less NoZ_NoUnpl_noHWE.out.log

```

WITHOUT RELATED SAMPLES

```bash

module purge
module load VCFtools/0.1.16-intel-2018b-Perl-5.28.0
module load BCFtools/1.9-intel-2018b

vcftools --gzvcf SPI_BJO_ROS_NU_SnpsAndIndels.AllNoW.raw.vcf.gz --remove-indv SPI002  --recode --recode-INFO-all --stdout | gzip > SPI_BJO_ROS_NoRelate_NU_SnpsAndIndels.AllNoW.raw.vcf.gz

mv out.log remove.log

ml purge
ml Miniconda3/4.9.2
conda activate miniconda3_envs/vcflib
gunzip SPI_BJO_ROS_NoRelate_NU_SnpsAndIndels.AllNoW.raw.vcf.gz
vcffixup SPI_BJO_ROS_NoRelate_NU_SnpsAndIndels.AllNoW.raw.vcf | gzip - > SPI_BJO_ROS_NoRelate_NU_SnpsAndIndels.AllNoW.filtered.raw.vcf.gz
gzip SPI_BJO_ROS_NoRelate_NU_SnpsAndIndels.AllNoW.raw.vcf.gz
conda deactivate

bcftools filter -e 'AC==0 || AC==AN' -O z -o SPI_BJO_ROS_NoRelate_NU_SnpsAndIndels.AllNoW.filtered2.raw.vcf.gz SPI_BJO_ROS_NoRelate_NU_SnpsAndIndels.AllNoW.filtered.raw.vcf.gz

#Total Variants
zcat SPI_BJO_ROS_NoRelate_NU_SnpsAndIndels.AllNoW.filtered2.raw.vcf.gz | grep -v "#" | wc -l

#Only SNPs 
module purge
module load VCFtools/0.1.16-intel-2018b-Perl-5.28.0
module load BCFtools/1.9-intel-2018b
vcftools --gzvcf SPI_BJO_ROS_NoRelate_NU_SnpsAndIndels.AllNoW.filtered2.raw.vcf.gz --remove-indels --out SPI_BJO_ROS_NoRelate_NU_Snps.AllNoW.raw

#Apply GATK filters
bcftools filter --SnpGap 10 -e 'QD < 2.0 || MQ < 40 || FS > 60.0 || SOR > 3 || MQRankSum < -12.5 || MQRankSum > 12.5 || ReadPosRankSum > 8.0 || ReadPosRankSum < -8.0' -O z -o SPI_BJO_ROS_NoRelate_NU_SnpsAndIndels.AllNoW.filtered1.vcf.gz SPI_BJO_ROS_NoRelate_NU_SnpsAndIndels.AllNoW.filtered2.raw.vcf.gz
bcftools view -H SPI_BJO_ROS_NoRelate_NU_SnpsAndIndels.AllNoW.filtered1.vcf.gz | wc -l 
bcftools view -v snps -H SPI_BJO_ROS_NoRelate_NU_SnpsAndIndels.AllNoW.filtered1.vcf.gz | wc -l

#the following step is kind of unneccessary because we are filtering for DP and GQ later
bcftools filter -S . -e 'FMT/DP<3 | FMT/GQ<15' -O z -o SPI_BJO_ROS_NoRelate_NU_SnpsAndIndels.AllNoW.filtered2.vcf.gz SPI_BJO_ROS_NoRelate_NU_SnpsAndIndels.AllNoW.filtered1.vcf.gz

ml purge
ml Miniconda3/4.9.2
conda activate miniconda3_envs/vcflib
gunzip SPI_BJO_ROS_NoRelate_NU_SnpsAndIndels.AllNoW.filtered2.vcf.gz
vcffixup SPI_BJO_ROS_NoRelate_NU_SnpsAndIndels.AllNoW.filtered2.vcf | gzip - > SPI_BJO_ROS_NoRelate_NU_SnpsAndIndels.AllNoW.filtered2.edit.vcf.gz
gzip SPI_BJO_ROS_NoRelate_NU_SnpsAndIndels.AllNoW.filtered2.vcf
conda deactivate

bcftools view -H SPI_BJO_ROS_NoRelate_NU_SnpsAndIndels.AllNoW.filtered2.edit.vcf.gz | wc -l 
bcftools view -v snps -H SPI_BJO_ROS_NoRelate_NU_SnpsAndIndels.AllNoW.filtered2.edit.vcf.gz | wc -l

# Exclude monomorphic SNPs and Indels and only include biallelic
vcftools --gzvcf SPI_BJO_ROS_NoRelate_NU_SnpsAndIndels.AllNoW.filtered2.edit.vcf.gz --remove-indels --min-alleles 2 --max-alleles 2 --maf 0.05 --max-maf 0.95 --recode --recode-INFO-all --stdout | gzip > SPI_BJO_ROS_NoRelate_NU_Snps.AllNoW.vcf.gz
zcat SPI_BJO_ROS_NoRelate_NU_Snps.AllNoW.vcf.gz | grep -v "#" | wc -l


vcftools --gzvcf SPI_BJO_ROS_NoRelate_NU_Snps.AllNoW.vcf.gz --not-chr Chr_Z --recode --recode-INFO-all --stdout | gzip > SPI_BJO_ROS_NoRelate_NU_Snps.Autosomes.vcf.gz

mv out.log NoZ.out.log
```

Filter More

```bash

module purge
module load VCFtools/0.1.16-intel-2018b-Perl-5.28.0
module load BCFtools/1.9-intel-2018b

#GLOBAL DEPTH
#maxDP 500
#minDP 250

bcftools filter -e 'INFO/DP[*] > 500 | INFO/DP[*] < 250 ' -O z -o SPI_BJO_ROS_NoRelate_NU_Snps.Autosomes.filtered1.vcf.gz SPI_BJO_ROS_NoRelate_NU_Snps.Autosomes.vcf.gz
zcat SPI_BJO_ROS_NoRelate_NU_Snps.Autosomes.filtered1.vcf.gz | grep -v "#" | wc -l

#OTHER FILTERS
#--max-missing 1
#--max-meanDP 43.4 
#--min-meanDP 10.9
#--minQ 30

vcftools --gzvcf SPI_BJO_ROS_NoRelate_NU_Snps.Autosomes.filtered1.vcf.gz --max-missing 1 --max-meanDP 43.4 --min-meanDP 10.9 --minQ 30 --remove-filtered-all --recode --recode-INFO-all --stdout | gzip > SPI_BJO_ROS_NoRelate_NU_Snps.Autosomes.filtered2.vcf.gz

mv out.log Autosomes.filtered2.log

less Autosomes.filtered2.log

vcftools --gzvcf SPI_BJO_ROS_NoRelate_NU_Snps.Autosomes.filtered2.vcf.gz --not-chr Chr_Unplaced --recode --recode-INFO-all --stdout | gzip > SPI_BJO_ROS_NoRelate_NU_Snps.Autosomes.filtered5.vcf.gz
mv out.log NoZ_NoUnpl_noHWW.out.log
less NoZ_NoUnpl_noHWW.out.log

```

NON VARIANTS

```bash


cp SPI_BJO_ROS_NU_SnpsAndIndels.*.raw.withnonvariant.vcf.gz* .
ls * | grep -v "Autosomes" | while read line ; do rm $line ; done

#Total Variants
zcat SPI_BJO_ROS_NU_SnpsAndIndels.Autosomes.raw.withnonvariant.vcf.gz | grep -v "#" | wc -l

module purge
module load VCFtools/0.1.16-intel-2018b-Perl-5.28.0
module load BCFtools/1.9-intel-2018b

for i in {01..24} ;
do echo ${i} ;
bcftools view -r Chr_${i} -Oz -o SPI_BJO_ROS_NU_SnpsAndIndels.Chr_${i}.raw.withnonvariant.vcf.gz SPI_BJO_ROS_NU_SnpsAndIndels.Autosomes.raw.withnonvariant.vcf.gz ;
done 

for i in {01..24} ;
do echo ${i} ;
sbatch FilterNonVariant.sh ${i} ;
done 

```

FilterNonVariant.sh

```bash

module load VCFtools/0.1.16-intel-2018b-Perl-5.28.0
module load BCFtools/1.9-intel-2018b

mkdir tmp.${1}
cd tmp.${1}

vcftools --gzvcf ../SPI_BJO_ROS_NU_SnpsAndIndels.Chr_${1}.raw.withnonvariant.vcf.gz --remove-indv SPI002  --recode --recode-INFO-all --stdout | gzip > SPI_BJO_ROS_NoRelate_NU_SnpsAndIndels.Chr_${1}.raw.withnonvariant.vcf.gz

ml purge
module load Miniconda3/4.9.2
export PS1=\$
source ${EBROOTMINICONDA3}/etc/profile.d/conda.sh
conda deactivate &>/dev/null
conda activate miniconda3_envs/vcflib

gunzip SPI_BJO_ROS_NoRelate_NU_SnpsAndIndels.Chr_${1}.raw.withnonvariant.vcf.gz
vcffixup SPI_BJO_ROS_NoRelate_NU_SnpsAndIndels.Chr_${1}.raw.withnonvariant.vcf | gzip - > SPI_BJO_ROS_NoRelate_NU_SnpsAndIndels.Chr_${1}.filtered.raw.withnonvariant.vcf.gz
gzip SPI_BJO_ROS_NoRelate_NU_SnpsAndIndels.Chr_${1}.raw.withnonvariant.vcf
conda deactivate

#Total Variants
zcat SPI_BJO_ROS_NoRelate_NU_SnpsAndIndels.Chr_${1}.filtered.raw.withnonvariant.vcf.gz | grep -v "#" | wc -l > Chr_${1}.count1

#Only SNPs 
module purge
module load VCFtools/0.1.16-intel-2018b-Perl-5.28.0
module load BCFtools/1.9-intel-2018b
vcftools --gzvcf SPI_BJO_ROS_NoRelate_NU_SnpsAndIndels.Chr_${1}.filtered.raw.withnonvariant.vcf.gz --remove-indels --out SPI_BJO_ROS_NoRelate_NU_Snps.Chr_${1}.raw.withnonvariant

#Apply GATK filters
gunzip SPI_BJO_ROS_NoRelate_NU_SnpsAndIndels.Chr_${1}.filtered.raw.withnonvariant.vcf.gz
bgzip SPI_BJO_ROS_NoRelate_NU_SnpsAndIndels.Chr_${1}.filtered.raw.withnonvariant.vcf
bcftools index SPI_BJO_ROS_NoRelate_NU_SnpsAndIndels.Chr_${1}.filtered.raw.withnonvariant.vcf.gz
bcftools filter --SnpGap 10 -e 'QD < 2.0 || MQ < 40 || FS > 60.0 || SOR > 3 || MQRankSum < -12.5 || MQRankSum > 12.5 || ReadPosRankSum > 8.0 || ReadPosRankSum < -8.0' -O z -o SPI_BJO_ROS_NoRelate_NU_SnpsAndIndels.Chr_${1}.withnonvariant.filtered1.vcf.gz SPI_BJO_ROS_NoRelate_NU_SnpsAndIndels.Chr_${1}.filtered.raw.withnonvariant.vcf.gz
bcftools view -H SPI_BJO_ROS_NoRelate_NU_SnpsAndIndels.Chr_${1}.withnonvariant.filtered1.vcf.gz | wc -l > Chr_${1}.count2
bcftools view -v snps -H SPI_BJO_ROS_NoRelate_NU_SnpsAndIndels.Chr_${1}.withnonvariant.filtered1.vcf.gz | wc -l >> Chr_${1}.count2

ml purge
ml Miniconda3/4.9.2
export PS1=\$
source ${EBROOTMINICONDA3}/etc/profile.d/conda.sh
conda deactivate &>/dev/null
conda activate miniconda3_envs/vcflib
gunzip SPI_BJO_ROS_NoRelate_NU_SnpsAndIndels.Chr_${1}.withnonvariant.filtered1.vcf.gz
vcffixup SPI_BJO_ROS_NoRelate_NU_SnpsAndIndels.Chr_${1}.withnonvariant.filtered1.vcf | gzip - > SPI_BJO_ROS_NoRelate_NU_SnpsAndIndels.Chr_${1}.withnonvariant.filtered1.edit.vcf.gz
gzip SPI_BJO_ROS_NoRelate_NU_SnpsAndIndels.Chr_${1}.withnonvariant.filtered1.vcf
conda deactivate

module purge
module load VCFtools/0.1.16-intel-2018b-Perl-5.28.0
module load BCFtools/1.9-intel-2018b

gunzip SPI_BJO_ROS_NoRelate_NU_SnpsAndIndels.Chr_${1}.withnonvariant.filtered1.edit.vcf.gz
bgzip SPI_BJO_ROS_NoRelate_NU_SnpsAndIndels.Chr_${1}.withnonvariant.filtered1.edit.vcf
bcftools index SPI_BJO_ROS_NoRelate_NU_SnpsAndIndels.Chr_${1}.withnonvariant.filtered1.edit.vcf.gz
bcftools view -H SPI_BJO_ROS_NoRelate_NU_SnpsAndIndels.Chr_${1}.withnonvariant.filtered1.edit.vcf.gz | wc -l > Chr_${1}.count3
bcftools view -v snps -H SPI_BJO_ROS_NoRelate_NU_SnpsAndIndels.Chr_${1}.withnonvariant.filtered1.edit.vcf.gz | wc -l >> Chr_${1}.count3

# Exclude monomorphic SNPs and Indels and only include biallelic
vcftools --gzvcf SPI_BJO_ROS_NoRelate_NU_SnpsAndIndels.Chr_${1}.withnonvariant.filtered1.edit.vcf.gz --remove-indels --max-alleles 2 --recode --recode-INFO-all --stdout | gzip > SPI_BJO_ROS_NoRelate_NU_Snps.Chr_${1}.withnonvariant.vcf.gz
zcat SPI_BJO_ROS_NoRelate_NU_Snps.Chr_${1}.withnonvariant.vcf.gz | grep -v "#" | wc -l > Chr_${1}.count4


```

Filter More

```bash

for i in {01..24} ;
do echo ${i} ;
sbatch FilterNonVariant2.sh ${i} ;
done 

```

FilterNonVariant2.sh

```bash

module load VCFtools/0.1.16-intel-2018b-Perl-5.28.0
module load BCFtools/1.9-intel-2018b

mkdir tmp.${1}
cd tmp.${1}

#GLOBAL DEPTH
#maxDP 500
#minDP 250

cp ../SPI_BJO_ROS_NoRelate_NU_Snps.Chr_${1}.withnonvariant.vcf.gz .
gunzip SPI_BJO_ROS_NoRelate_NU_Snps.Chr_${1}.withnonvariant.vcf.gz
bgzip SPI_BJO_ROS_NoRelate_NU_Snps.Chr_${1}.withnonvariant.vcf
bcftools index SPI_BJO_ROS_NoRelate_NU_Snps.Chr_${1}.withnonvariant.vcf.gz
bcftools filter -e 'INFO/DP[*] > 500 | INFO/DP[*] < 250 ' -O z -o SPI_BJO_ROS_NoRelate_NU_Snps.Chr_${1}.withnonvariant.filtered1.vcf.gz SPI_BJO_ROS_NoRelate_NU_Snps.Chr_${1}.withnonvariant.vcf.gz
zcat SPI_BJO_ROS_NoRelate_NU_Snps.Chr_${1}.withnonvariant.filtered1.vcf.gz | grep -v "#" | wc -l > Chr_${1}.count5
rm SPI_BJO_ROS_NoRelate_NU_Snps.Chr_${1}.withnonvariant.vcf.gz

#OTHER FILTERS
#--max-missing 1
#--max-meanDP 43.4 
#--min-meanDP 10.9
#--minQ 30 - cannot do minQ because invariant have 0

vcftools --gzvcf SPI_BJO_ROS_NoRelate_NU_Snps.Chr_${1}.withnonvariant.filtered1.vcf.gz --max-missing 1 --max-meanDP 43.4 --min-meanDP 10.9 --remove-filtered-all --recode --recode-INFO-all --stdout | gzip > SPI_BJO_ROS_NoRelate_NU_Snps.Chr_${1}.withnonvariant.filtered2.vcf.gz

mv SPI_BJO_ROS_NoRelate_NU_Snps.Chr_${1}.withnonvariant.filtered2.vcf.gz SPI_BJO_ROS_NoRelate_NU_Snps.Chr_${1}.withnonvariant.filtered3.vcf.gz
gunzip SPI_BJO_ROS_NoRelate_NU_Snps.Chr_${1}.withnonvariant.filtered3.vcf.gz
bgzip SPI_BJO_ROS_NoRelate_NU_Snps.Chr_${1}.withnonvariant.filtered3.vcf
bcftools index SPI_BJO_ROS_NoRelate_NU_Snps.Chr_${1}.withnonvariant.filtered3.vcf.gz

```

merge the files

```bash

module purge
module load VCFtools/0.1.16-intel-2018b-Perl-5.28.0
module load BCFtools/1.9-intel-2018b

for i in {01..24} ;
do echo ${i} ;
gunzip SPI_BJO_ROS_NoRelate_NU_Snps.Chr_${i}.withnonvariant.filtered3.vcf.gz
bgzip SPI_BJO_ROS_NoRelate_NU_Snps.Chr_${i}.withnonvariant.filtered3.vcf
bcftools index SPI_BJO_ROS_NoRelate_NU_Snps.Chr_${i}.withnonvariant.filtered3.vcf.gz
rm SPI_BJO_ROS_NoRelate_NU_Snps.Chr_${i}.withnonvariant.filtered1.vcf.gz
done

ls SPI_BJO_ROS_NoRelate_NU_Snps.Chr_*.withnonvariant.filtered3.vcf.gz > samplelist

bcftools concat -n -f samplelist -O b -o SPI_BJO_ROS_NoRelate_NU_Snps.Autosomes.withnonvariant.filtered3.vcf.gz 

```