# LD pruning

Prune SNPs based on LD

```bash

module purge
module load VCFtools/0.1.16-intel-2018b-Perl-5.28.0
module load BCFtools/1.9-intel-2018b

#Get chromosome map for PLINK
bcftools view -H SPI_BJO_ROS_NU_Snps.Autosomes.filtered4.vcf.gz | cut -f 1 | uniq | awk '{print $0"\t"$0}' > SPI_BJO_ROS_NU_Snps.Autosomes.filtered4.chrom-map.txt

# Convert to PLINK
vcftools --gzvcf SPI_BJO_ROS_NU_Snps.Autosomes.filtered4.vcf.gz --plink --chrom-map SPI_BJO_ROS_NU_Snps.Autosomes.filtered4.chrom-map.txt --out SPI_BJO_ROS_NU_Snps.Autosomes.filtered4.vcf

#Update sample info and add colonies
cat SPI_BJO_ROS_NU_Snps.Autosomes.filtered4.vcf.ped | cut -f1 > tmp
cat SPI_BJO_ROS_NU_Snps.Autosomes.filtered4.vcf.ped | cut -c 1,2,3 | paste tmp - > tmp2
awk '{print $1"\t"$1"\t"$2"\t"$1}' tmp2 > updateID.txt
rm -rf tmp*

module purge
module load PLINK/1.9b_6.13-x86_64
plink -file SPI_BJO_ROS_NU_Snps.Autosomes.filtered4.vcf -aec -update-ids updateID.txt -make-bed -recode -out SPI_BJO_ROS_NU_Snps.Autosomes.plink

# Identify LD with Plink (check Mark or evomics) - use 55GB RAM
# window size kb = 1
# shift between each window (SNPs) = 1 (all SNPs within 1kb are compared)

#LD plots
plink -bfile SPI_BJO_ROS_NU_Snps.Autosomes.plink -aec --r2 --ld-window-kb 50 --ld-window-r2 0 -out SPI_BJO_ROS_NU_Snps.Autosomes.plink.LD
gzip SPI_BJO_ROS_NU_Snps.Autosomes.plink.LD.ld

# r2 = 0.2 (SNPs greater that the value are removed)

plink -bfile SPI_BJO_ROS_NU_Snps.Autosomes.plink -aec -indep-pairwise 50 1 0.2 -out SPI_BJO_ROS_NU_Snps.Autosomes.plink.ldSnps
plink -bfile SPI_BJO_ROS_NU_Snps.Autosomes.plink -aec -exclude SPI_BJO_ROS_NU_Snps.Autosomes.plink.ldSnps.prune.out -out SPI_BJO_ROS_NU_Snps.Autosomes.plink.pruned -make-bed

module purge
module load VCFtools/0.1.16-intel-2018b-Perl-5.28.0
module load BCFtools/1.9-intel-2018b

#then select only the LD.prune.in.intervals with bcftools to create vcf 
cp SPI_BJO_ROS_NU_Snps.Autosomes.plink.ldSnps.prune.in SPI_BJO_ROS_NU_Snps.Autosomes.plink.ldSnps.prune.in.list
sed -i 's/:/\t/g' SPI_BJO_ROS_NU_Snps.Autosomes.plink.ldSnps.prune.in.list
gunzip SPI_BJO_ROS_NU_Snps.Autosomes.filtered4.vcf.gz
bgzip SPI_BJO_ROS_NU_Snps.Autosomes.filtered4.vcf
bcftools index SPI_BJO_ROS_NU_Snps.Autosomes.filtered4.vcf.gz

bcftools view -O z -R SPI_BJO_ROS_NU_Snps.Autosomes.plink.ldSnps.prune.in.list -o SPI_BJO_ROS_NU_Snps.Autosomes.filtered5.vcf.gz SPI_BJO_ROS_NU_Snps.Autosomes.filtered4.vcf.gz

zcat SPI_BJO_ROS_NU_Snps.Autosomes.filtered5.vcf.gz | grep -v "#" | wc -l

#sort vcf to have snps in correct order
zcat SPI_BJO_ROS_NU_Snps.Autosomes.filtered5.vcf.gz | grep "#" > headers
zcat SPI_BJO_ROS_NU_Snps.Autosomes.filtered5.vcf.gz | grep -v "#" | sort -k1,1V -k2,2n | cat headers - > SPI_BJO_ROS_NU_Snps.Autosomes.filtered.pruned.sorted.vcf
gzip SPI_BJO_ROS_NU_Snps.Autosomes.filtered.pruned.sorted.vcf
zcat SPI_BJO_ROS_NU_Snps.Autosomes.filtered.pruned.sorted.vcf.gz | grep -v "#" | wc -l

```

Check relatedness

```bash

# Close relatives can be inferred fairly reliably based on the estimated kinship coefficients as shown 
# in the following simple algorithm: an estimated kinship coefficient range >0.354, [0.177, 0.354], 
# [0.0884, 0.177] and [0.0442, 0.0884] corresponds to duplicate/MZ twin, 1st-degree, 2nd-degree, and
# 3rd-degree relationships respectively.

#VCFTools
module purge
module load VCFtools/0.1.16-intel-2018b-Perl-5.28.0
module load BCFtools/1.9-intel-2018b

# The --relatedness method based on doi:10.1038/ng.608 is only appropriate for large cohorts. 
# Although this is not properly indicated in the documentation, it is revealed in questions on 
# SourceForge. Looking at the paper equation (6) in supplementary, the values of AF near 0 or 1 
# make the equation numerically unstable. I would just ignore this result.
# The --relatedness2 method is based on the KING inference, and you have a very good tutorial here. 
# You can interpret the relatedness_phi as the probability to find identical alleles when randomly 
# sampling one allele from each heterozygous individual. So for one individual AB, and the parent AC,
# there is p=0.25 to choose A from both individuals. That probability is 0.5 when AB is compared to AB.

vcftools --gzvcf SPI_BJO_ROS_NU_Snps.Autosomes.filtered.pruned.sorted.vcf.gz --relatedness
vcftools --gzvcf SPI_BJO_ROS_NU_Snps.Autosomes.filtered.pruned.sorted.vcf.gz --relatedness2
mv out.relatedness SPI_BJO_ROS_NU_Snps.Autosomes.filtered.pruned.sorted.vcf.relatedness
mv out.relatedness2 SPI_BJO_ROS_NU_Snps.Autosomes.filtered.pruned.sorted.vcf.relatedness2

```

WithoutRelated

```bash

mkdir NoRelate
cd NoRelate/

cp ../../VCF_Filter/NoRelate/SPI_BJO_ROS_NoRelate_NU_Snps.Autosomes.filtered4.vcf.gz .

module purge
module load VCFtools/0.1.16-intel-2018b-Perl-5.28.0
module load BCFtools/1.9-intel-2018b

#Get chromosome map for PLINK
bcftools view -H SPI_BJO_ROS_NoRelate_NU_Snps.Autosomes.filtered4.vcf.gz | cut -f 1 | uniq | awk '{print $0"\t"$0}' > SPI_BJO_ROS_NoRelate_NU_Snps.Autosomes.filtered4.chrom-map.txt

# Convert to PLINK
vcftools --gzvcf SPI_BJO_ROS_NoRelate_NU_Snps.Autosomes.filtered4.vcf.gz --plink --chrom-map SPI_BJO_ROS_NoRelate_NU_Snps.Autosomes.filtered4.chrom-map.txt --out SPI_BJO_ROS_NoRelate_NU_Snps.Autosomes.filtered4.vcf

#Update sample info and add colonies
cat SPI_BJO_ROS_NoRelate_NU_Snps.Autosomes.filtered4.vcf.ped | cut -f1 > tmp
cat SPI_BJO_ROS_NoRelate_NU_Snps.Autosomes.filtered4.vcf.ped | cut -c 1,2,3 | paste tmp - > tmp2
awk '{print $1"\t"$1"\t"$2"\t"$1}' tmp2 > updateID.txt
rm -rf tmp*

module purge
module load PLINK/1.9b_6.13-x86_64
plink -file SPI_BJO_ROS_NoRelate_NU_Snps.Autosomes.filtered4.vcf -aec -update-ids updateID.txt -make-bed -recode -out SPI_BJO_ROS_NoRelate_NU_Snps.Autosomes.plink

# Identify LD with Plink (check Mark or evomics) - use 55GB RAM
# window size kb = 1
# shift between each window (SNPs) = 1 (all SNPs within 1kb are compared)

#LD plots
plink -bfile SPI_BJO_ROS_NoRelate_NU_Snps.Autosomes.plink -aec --r2 gz --thin 0.05 --ld-window-kb 50 --ld-window 10000 --ld-window-r2 0 -out SPI_BJO_ROS_NoRelate_NU_Snps.Autosomes.plink.LD

#LOCAL
mkdir /Users/oliverke/Desktop/PhD_Bioinformatics/Mapping_NewReference/nuclear/GATK/LD/NoRelated
cd /Users/oliverke/Desktop/PhD_Bioinformatics/Mapping_NewReference/nuclear/GATK/LD/NoRelated
scp oliverke@saga.sigma2.no:/cluster/work/users/oliverke/HighCov/NucAnalysis/GATK/LD/NoRelate/SPI_BJO_ROS_NoRelate_NU_Snps.Autosomes.plink.LD.ld.gz .
gunzip SPI_BJO_ROS_NoRelate_NU_Snps.Autosomes.plink.LD.ld.gz

# r2 = 0.2 (SNPs greater that the value are removed)
plink -bfile SPI_BJO_ROS_NoRelate_NU_Snps.Autosomes.plink -aec -indep-pairwise 50 1 0.2 -out SPI_BJO_ROS_NoRelate_NU_Snps.Autosomes.plink.ldSnps
plink -bfile SPI_BJO_ROS_NoRelate_NU_Snps.Autosomes.plink -aec -exclude SPI_BJO_ROS_NoRelate_NU_Snps.Autosomes.plink.ldSnps.prune.out -out SPI_BJO_ROS_NoRelate_NU_Snps.Autosomes.plink.pruned -make-bed

module purge
module load VCFtools/0.1.16-intel-2018b-Perl-5.28.0
module load BCFtools/1.9-intel-2018b

#then select only the LD.prune.in.intervals with bcftools to create vcf 
cp SPI_BJO_ROS_NoRelate_NU_Snps.Autosomes.plink.ldSnps.prune.in SPI_BJO_ROS_NoRelate_NU_Snps.Autosomes.plink.ldSnps.prune.in.list
sed -i 's/:/\t/g' SPI_BJO_ROS_NoRelate_NU_Snps.Autosomes.plink.ldSnps.prune.in.list
gunzip SPI_BJO_ROS_NoRelate_NU_Snps.Autosomes.filtered4.vcf.gz
bgzip SPI_BJO_ROS_NoRelate_NU_Snps.Autosomes.filtered4.vcf
bcftools index SPI_BJO_ROS_NoRelate_NU_Snps.Autosomes.filtered4.vcf.gz

bcftools view -O z -R SPI_BJO_ROS_NoRelate_NU_Snps.Autosomes.plink.ldSnps.prune.in.list -o SPI_BJO_ROS_NoRelate_NU_Snps.Autosomes.filtered5.vcf.gz SPI_BJO_ROS_NoRelate_NU_Snps.Autosomes.filtered4.vcf.gz

zcat SPI_BJO_ROS_NoRelate_NU_Snps.Autosomes.filtered5.vcf.gz | grep -v "#" | wc -l

#sort vcf to have snps in correct order
zcat SPI_BJO_ROS_NoRelate_NU_Snps.Autosomes.filtered5.vcf.gz | grep "#" > headers
zcat SPI_BJO_ROS_NoRelate_NU_Snps.Autosomes.filtered5.vcf.gz | grep -v "#" | sort -k1,1V -k2,2n | cat headers - > SPI_BJO_ROS_NoRelate_NU_Snps.Autosomes.filtered.pruned.sorted.vcf
gzip SPI_BJO_ROS_NoRelate_NU_Snps.Autosomes.filtered.pruned.sorted.vcf
zcat SPI_BJO_ROS_NoRelate_NU_Snps.Autosomes.filtered.pruned.sorted.vcf.gz | grep -v "#" | wc -l


```