# SmartPCA (NoZ/Unpl)

Convert files for smartPCA

```bash


cp ../LD/SPI_BJO_ROS_NU_Snps.Autosomes.filtered.pruned.sorted.vcf.* .

# edit map file to only have chromosome numbers and to have a valid SNP identifier
cat SPI_BJO_ROS_NU_Snps.Autosomes.filtered.pruned.sorted.vcf.map | sed 's/Chr_//g' > SPI_BJO_ROS_NU_Snps.Autosomes.filtered.pruned.sorted.vcf.edit.map
awk '{print $1"\t"$1":"$4"\t"$3"\t"$4}' SPI_BJO_ROS_NU_Snps.Autosomes.filtered.pruned.sorted.vcf.edit.map > SPI_BJO_ROS_NU_Snps.Autosomes.filtered.pruned.sorted.vcf.edit2.map

# update population information for each Individ. and save as the .pedind file
cat SPI_BJO_ROS_NU_Snps.Autosomes.filtered.pruned.sorted.vcf.ped | cut -d ' ' -f1-6 > tmp
cat SPI_BJO_ROS_NU_Snps.Autosomes.filtered.pruned.sorted.vcf.ped | cut -c1-3 | paste tmp - > SPI_BJO_ROS_NU_Snps.Autosomes.filtered.pruned.sorted.vcf.pedind
rm -rf tmp

```

On Local

```bash


# get add. files for next commands
IndToMeta.txt
PCAfromSmartPCA.R .
PuffinNU_PCA* .

# make sure IndToMeta and pedind have same samples and same order
awk '{print $1}' SPI_BJO_ROS_NU_Snps.Autosomes.filtered.pruned.sorted.vcf.pedind | while read line ;
do grep -w ${line} IndToMeta.txt >> IndToMeta.edit.txt ;
done

#change pedind so it gets read by smartpca
awk -F " " '{print "1""\t"$2"\t"$3"\t"$4"\t"$5"\t"$7}' SPI_BJO_ROS_NU_Snps.Autosomes.filtered.pruned.sorted.vcf.pedind | paste - IndToMeta.edit.txt | awk -F "\t" '{print $1"\t"$2"\t"$3"\t"$4"\t"$9"\t"$6}' > SPI_BJO_ROS_NU_Snps.Autosomes.filtered.pruned.sorted.vcf.edit.pedind

# edit filenames in par files

conda activate eigensoft
smartpca -p PuffinNU_PCA_withoutlier.par > PuffinNU_PCALog_withOutlier.out
conda deactivate

# then PCAfromSmartPCA.R

```

No related samples

```bash

cp ../../LD/NoRelate/SPI_BJO_ROS_NoRelate_NU_Snps.Autosomes.filtered.pruned.sorted.vcf.* .

# edit map file to only have chromosome numbers and to have a valid SNP identifier
cat SPI_BJO_ROS_NoRelate_NU_Snps.Autosomes.filtered.pruned.sorted.vcf.map | sed 's/Chr_//g' > SPI_BJO_ROS_NoRelate_NU_Snps.Autosomes.filtered.pruned.sorted.vcf.edit.map
awk '{print $1"\t"$1":"$4"\t"$3"\t"$4}' SPI_BJO_ROS_NoRelate_NU_Snps.Autosomes.filtered.pruned.sorted.vcf.edit.map > SPI_BJO_ROS_NoRelate_NU_Snps.Autosomes.filtered.pruned.sorted.vcf.edit2.map

# update population information for each Individ. and save as the .pedind file
cat SPI_BJO_ROS_NoRelate_NU_Snps.Autosomes.filtered.pruned.sorted.vcf.ped | cut -d ' ' -f1-6 > tmp
cat SPI_BJO_ROS_NoRelate_NU_Snps.Autosomes.filtered.pruned.sorted.vcf.ped | cut -c1-3 | paste tmp - > SPI_BJO_ROS_NoRelate_NU_Snps.Autosomes.filtered.pruned.sorted.vcf.pedind
rm -rf tmp

#LOCAL

cp ../PuffinNU_PCA_withoutlier.par .
cp ../IndToMeta.edit.txt .
cp ../PCAfromSmartPCA.R .

# make sure IndToMeta and pedind have same samples and same order
touch IndToMeta.edit2.txt
awk '{print $1}' SPI_BJO_ROS_NoRelate_NU_Snps.Autosomes.filtered.pruned.sorted.vcf.pedind | while read line ;
do grep -w ${line} IndToMeta.edit.txt >> IndToMeta.edit2.txt ;
done

#change pedind so it gets read by smartpca
awk -F " " '{print "1""\t"$2"\t"$3"\t"$4"\t"$5"\t"$7}' SPI_BJO_ROS_NoRelate_NU_Snps.Autosomes.filtered.pruned.sorted.vcf.pedind | paste - IndToMeta.edit2.txt | awk -F "\t" '{print $1"\t"$2"\t"$3"\t"$4"\t"$9"\t"$6}' > SPI_BJO_ROS_NoRelate_NU_Snps.Autosomes.filtered.pruned.sorted.vcf.edit.pedind

# edit filenames in par files

conda activate eigensoft
smartpca -p PuffinNU_PCA_withoutlier.par > PuffinNU_PCALog_withOutlier.out
conda deactivate

# then PCAfromSmartPCA.R

```
