# Admixture

```bash

cp ../LD/NoRelate/SPI_BJO_ROS_NoRelate_NU_Snps.Autosomes.filtered.pruned.sorted.vcf* 

#make Plink files
module purge
module load PLINK/1.9b_6.13-x86_64
cat SPI_BJO_ROS_NoRelate_NU_Snps.Autosomes.filtered.pruned.sorted.vcf.map | sed 's/Chr_//g' > SPI_BJO_ROS_NoRelate_NU_Snps.Autosomes.filtered.pruned.sorted.vcf.edit.map
zcat SPI_BJO_ROS_NoRelate_NU_Snps.Autosomes.filtered.pruned.sorted.vcf.gz | sed 's/Chr_//g' - > SPI_BJO_ROS_NoRelate_NU_Snps.Autosomes.filtered.pruned.sorted.edit.vcf
gzip SPI_BJO_ROS_NoRelate_NU_Snps.Autosomes.filtered.pruned.sorted.edit.vcf
plink --vcf SPI_BJO_ROS_NoRelate_NU_Snps.Autosomes.filtered.pruned.sorted.edit.vcf.gz --aec --recode --make-bed --out SPI_BJO_ROS_NoRelate_NU_Snps.Autosomes.filtered.pruned.sorted.vcf.edit2

#Run Admixture
for K in $(seq 1 6) ;
do echo ${K};
for i in $(seq 1 50) ;
do sbatch Admixture.sh SPI_BJO_ROS_NoRelate_NU_Snps.Autosomes.filtered.pruned.sorted.vcf.edit2 ${K} ${i};
done ;
done ;

####### Admixture.sh #######

module purge
module load ADMIXTURE/1.3.0

echo ${1}
echo ${2}
echo ${3}

mkdir temp_${1}_${2}_${3}
cd temp_${1}_${2}_${3}

#making a random seed
R=$(echo $RANDOM % 100 | bc)
S=$(bc <<<"scale=0; ${R} * ${2} * ${3}")

echo "Run Admixture"
echo "admixture ${1}.bed k=${2} replicate=${3}"

admixture --cv=10 -s ${S} ../${1}.bed ${2} -j4 | tee ${1}_${2}_${3}.out

mv ${1}.${2}.P ${1}.${2}.${3}.P
mv ${1}.${2}.Q ${1}.${2}.${3}.Q

mv ${1}.${2}.${3}.* ../
mv ${1}_${2}_${3}.out ../

cd ../

rm -r temp_${1}_${2}_${3}

```

```bash

#log likelihoods for best K
(for log in `ls *.out`; do grep '^Loglikelihood:' $log | awk -F ':' '{print $2}' ; done) > logfile

#Zip file for Clumpak run
for i in 1 2 3 4 5 6 ;
do zip K${i}_SPI_BJO_ROS_NoRelate_NU_Snps.Autosomes.filtered.pruned.sorted.zip SPI_BJO_ROS_NoRelate_NU_Snps.Autosomes.filtered.pruned.sorted.vcf.edit2.${i}.*.Q ;
done ;
zip SPI_BJO_ROS_NoRelate_NU_Snps.Autosomes.filtered.pruned.sorted.zip K*_SPI_BJO_ROS_NoRelate_NU_Snps.Autosomes.filtered.pruned.sorted.zip


#use NgsAdmix2Clumpak.R

#sort file
sort -n -k1,1 logfile_formatted.txt | awk -F ' ' '{print $1"\t"$2}' > logfile_formatted_clumpak.txt

# then clumpak ; http://clumpak.tau.ac.il/bestK.html
# with logfile and zip file
# save output zipfolders as EvannoOut and ClumpakOut

#check best K's (Evanno) and select mean Qmatrix file based Clumpakk

awk '{print $6"\t"$7}' Clumpak/1639561987/K=2/MajorCluster/CLUMPP.files/ClumppIndFile.output > K2_Clumpak_MajorQmatrix.txt
#awk '{print $6"\t"$7"\t"$8}' Clumpak/1621093048/K=3/MajorCluster/CLUMPP.files/ClumppIndFile.output > K3_Clumpak_MajorQmatrix.txt
#awk '{print $6"\t"$7"\t"$8"\t"$9}' Clumpak/1621093048/K=4/MajorCluster/CLUMPP.files/ClumppIndFile.output > K4_Clumpak_MajorQmatrix.txt

# then Admixture_GenoLike.R for Plot

```

Check best Admixture

```bash

(for log in `ls *edit2_2_*.out`; do grep '^Loglikelihood:' $log | awk -F ':' '{print $2}' ; done) > logfile_tmp1
(for log in `ls *edit2_2_*.out`; do echo $log; done) > logfile_tmp2
paste logfile_tmp1 logfile_tmp2 > logfile_k2
rm logfile_tmp1
rm logfile_tmp2
sort -n -r -k1,1 logfile_k2 | head  > FilePrefix_k2

cat SPI_BJO_ROS_NoRelate_NU_Snps.Autosomes.filtered.pruned.sorted.vcf.edit2.2.9.Q
cat SPI_BJO_ROS_NoRelate_NU_Snps.Autosomes.filtered.pruned.sorted.vcf.edit2.2.26.Q

#approx. the same as CLUMPAK

```

F3 stats

```bash

mkdir Treemix
cd Treemix

cp ../SPI_BJO_ROS_NoRelate_NU_Snps.Autosomes.filtered.pruned.sorted.vcf.gz .

module purge
module load VCFtools/0.1.16-intel-2018b-Perl-5.28.0
module load BCFtools/1.9-intel-2018b

bcftools query -l SPI_BJO_ROS_NoRelate_NU_Snps.Autosomes.filtered.pruned.sorted.vcf.gz . | awk '{print $1"\t"$1"\t"substr($1,1,3)}' > puffins.clust

module purge
module load PLINK/1.9b_6.13-x86_64
module load VCFtools/0.1.16-intel-2018b-Perl-5.28.0
module load Python/2.7.15-foss-2018b

chmod +x vcf2treemix.sh

./vcf2treemix.sh SPI_BJO_ROS_NoRelate_NU_Snps.Autosomes.filtered.pruned.sorted.vcf.gz puffins.clust

```

vcf2treemix.sh

```bash

#!/bin/bash

# Script to convert vcf file to Treemix format
# by Joana Meier
# requires vcftools, plink and plink2treemix.py from Treemix package to be installed
# takes as input the vcf file and the clust file

if [ $# -ne 2 ]
 then
 echo "Please provide the following arguments: <vcf file> <clust file>"
 echo "The .clust file contains three columns: samplename\tsamplename\tgroup"
 exit 1
fi

clust=$2
file=${1%.gz}
file=${file%.vcf}

# Use VCFtools to make a map and a ped file, using only bi-allelic SNPs with mac 2 (also creates a log file)
if [ -s $file.vcf.gz ]
then

 # Get a .map and .ped file
 vcftools --gzvcf $file".vcf.gz" \
         --plink --mac 2 --remove-indels --max-alleles 2 \
         --out $file

else
 file=${file%.vcf}
 vcftools --vcf $file".vcf" \
         --plink --mac 2 --remove-indels --max-alleles 2  \
         --out $file

fi

# Adjust the map file to allow for non-human chromosome names (else the new plink version struggles)
awk -F"\t" '{
        split($2,chr,":")
	$1="1"
	$2="1:"chr[2]
        print $0
}' ${file}.map > better.map
mv better.map ${file}.map

# convert it to a stratified frq file, also creates .bed, .bim, .fam, .log, .nosex
plink --file $file --make-bed --out $file --allow-no-sex --allow-extra-chr 0
plink --bfile $file --freq --missing --within $2 --out $file --allow-no-sex --allow-extra-chr 0

# zip it
gzip $file".frq.strat"

# create input file for Treemix
python plink2treemix.py $file".frq.strat.gz" $file".treemix.frq.gz"

# unzip allele frequency information
gunzip $file".treemix.frq.gz"
gunzip $file".frq.strat.gz"

# make a file with the positions
awk 'BEGIN{print "scaffold_pos\tscaffold\tpos"}{split($2,pos,":");print $2"\t"pos[1]"\t"pos[2]}' $file".map" > $file".positions"
paste $file".positions" $file".treemix.frq" > $file".frequencies"

awk '{printf $0
	for(i = 4; i <= NF; i++){
		split($i,values,",")
		if((values[1]+values[2])>0) freq=values[1]/(values[1]+values[2])
		else freq=0
		printf freq"\t"
	}
	printf "\n"}' $file".frequencies" > $file".frequencies2"
mv $file".frequencies2" $file".frequencies"

awk 'BEGIN{scaffold="";pos=0;newpos=0}
	{if($2==scaffold){newpos=pos+$3}else{scaffold=$2;pos=newpos};chpos=pos+$3;print $0,chpos}' \
	$file".frequencies" > $file".frequencies.newpos"

gzip $file".treemix.frq"

```

plink2treemix.py

```bash

#!/usr/bin/python

import sys, os, gzip

if len(sys.argv) < 3:
	print "plink2treemix.py [gzipped input file] [gzipped output file]"
	print "ERROR: improper command line"
	exit(1)
infile = gzip.open(sys.argv[1])
outfile = gzip.open(sys.argv[2], "w")

pop2rs = dict()
rss = list()
rss2 = set()

line = infile.readline()
line = infile.readline()
while line:
    line = line.strip().split()
    rs = line[1]
    pop = line[2]
    mc = line[6]
    total = line[7]
    if rs not in rss2: 
    	rss.append(rs)
    rss2.add(rs)
    if pop2rs.has_key(pop)==0:
        pop2rs[pop] =  dict()
    if pop2rs[pop].has_key(rs)==0:
        pop2rs[pop][rs] = " ".join([mc, total])
    line = infile.readline()

pops = pop2rs.keys()
for pop in pops:
    print >> outfile, pop,
print >> outfile, ""

for rs in rss:
    for pop in pops:
        tmp = pop2rs[pop][rs].split()
        c1 = int(tmp[0])
        c2 = int(tmp[1])
        c3 = c2-c1
        print >> outfile, ",".join([str(c1), str(c3)]),
    print >> outfile, ""
```

```bash

sbatch F3.sh

mkdir temp.f3

cp SPI_BJO_ROS_NoRelate_NU_Snps.Autosomes.filtered.pruned.sorted.treemix.frq.gz temp.f3/

cd temp.f3/

## Load modules
module load TreeMix/1.13-intel-2018b

## Run treemix
threepop -i SPI_BJO_ROS_NoRelate_NU_Snps.Autosomes.filtered.pruned.sorted.treemix.frq.gz -k 500 > SPI_BJO_ROS_NoRelate_NU_Snps.Autosomes.filtered.pruned.sorted.treemix.frq.F3stat

rm SPI_BJO_ROS_NoRelate_NU_Snps.Autosomes.filtered.pruned.sorted.treemix.frq.gz
cp * ../
cd ../
rm -r temp.f3/

```