# Fst, Tajima and Nuc Diversity Sliding Window


FST

```bash

cp ../VCF_Filter/NoRelate/SPI_BJO_ROS_NoRelate_NU_Snps.Autosomes.filtered4.vcf.gz .

module purge
module load VCFtools/0.1.16-intel-2018b-Perl-5.28.0
module load BCFtools/1.9-intel-2018b

# check individuals
bcftools query -l SPI_BJO_ROS_NoRelate_NU_Snps.Autosomes.filtered4.vcf.gz
# how many individuals
bcftools query -l SPI_BJO_ROS_NoRelate_NU_Snps.Autosomes.filtered4.vcf.gz | wc -l
# how many variants?
gunzip SPI_BJO_ROS_NoRelate_NU_Snps.Autosomes.filtered4.vcf.gz
bgzip SPI_BJO_ROS_NoRelate_NU_Snps.Autosomes.filtered4.vcf
bcftools index SPI_BJO_ROS_NoRelate_NU_Snps.Autosomes.filtered4.vcf.gz
bcftools index -n SPI_BJO_ROS_NoRelate_NU_Snps.Autosomes.filtered4.vcf.gz #speeds up counting SNPs DRASTICALLY!

# generate all the sample names
bcftools query -l SPI_BJO_ROS_NoRelate_NU_Snps.Autosomes.filtered4.vcf.gz > samples
# get the samples per population
grep "BJO" samples > bjornoya
grep "ROS" samples > rost
grep "SPI" samples > spitsbergen

# However, the population file also needs a column with the population denoted in it. So we can generate this with awk.
awk '{print $1"\tBJO"}' bjornoya > pop_file
awk '{print $1"\tROS"}' rost >> pop_file
awk '{print $1"\tSPI"}' spitsbergen >> pop_file

#GLOBAL FST

vcftools --gzvcf SPI_BJO_ROS_NoRelate_NU_Snps.Autosomes.filtered4.vcf.gz \
--weir-fst-pop rost \
--weir-fst-pop spitsbergen \
--out ./ros_spi

vcftools --gzvcf SPI_BJO_ROS_NoRelate_NU_Snps.Autosomes.filtered4.vcf.gz \
--weir-fst-pop rost \
--weir-fst-pop bjornoya \
--out ./ros_bjo

vcftools --gzvcf SPI_BJO_ROS_NoRelate_NU_Snps.Autosomes.filtered4.vcf.gz \
--weir-fst-pop bjornoya \
--weir-fst-pop spitsbergen \
--out ./bjo_spi

#SLIDING WINDOW
#50 Kb sliding windows with a 25 Kb
vcftools --gzvcf SPI_BJO_ROS_NoRelate_NU_Snps.Autosomes.filtered4.vcf.gz \
--weir-fst-pop rost \
--weir-fst-pop spitsbergen \
--fst-window-size 50000 --fst-window-step 25000 \
--out ./ros_spi_sliding

vcftools --gzvcf SPI_BJO_ROS_NoRelate_NU_Snps.Autosomes.filtered4.vcf.gz \
--weir-fst-pop rost \
--weir-fst-pop bjornoya \
--fst-window-size 50000 --fst-window-step 25000 \
--out ./ros_bjo_sliding

vcftools --gzvcf SPI_BJO_ROS_NoRelate_NU_Snps.Autosomes.filtered4.vcf.gz \
--weir-fst-pop bjornoya \
--weir-fst-pop spitsbergen \
--fst-window-size 50000 --fst-window-step 25000 \
--out ./bjo_spi_sliding

```


 Tajima, pi and dxy - Prep

```bash

cp /cluster/work/users/oliverke/HighCov/NucAnalysis/GATK/VCF_Filter/NoRelate/NonVariants/SPI_BJO_ROS_NoRelate_NU_Snps.Autosomes.withnonvariant.filtered3.vcf.gz .

module purge
module load VCFtools/0.1.16-intel-2018b-Perl-5.28.0
module load BCFtools/1.9-intel-2018b

bcftools index SPI_BJO_ROS_NoRelate_NU_Snps.Autosomes.withnonvariant.filtered3.vcf.gz

bcftools index -n SPI_BJO_ROS_NoRelate_NU_Snps.Autosomes.withnonvariant.filtered3.vcf.gz

# generate all the sample names
bcftools query -l SPI_BJO_ROS_NoRelate_NU_Snps.Autosomes.withnonvariant.filtered3.vcf.gz > samples
# get the samples per population
grep "BJO" samples > bjornoya
grep "ROS" samples > rost
grep "SPI" samples > spitsbergen

# However, the population file also needs a column with the population denoted in it. So we can generate this with awk.
awk '{print $1"\tBJO"}' bjornoya > pop_file
awk '{print $1"\tROS"}' rost >> pop_file
awk '{print $1"\tSPI"}' spitsbergen >> pop_file


bcftools view -S bjornoya -O z -o BJO_NoRelate_NU_Snps.Autosomes.withnonvariant.filtered3.vcf.gz SPI_BJO_ROS_NoRelate_NU_Snps.Autosomes.withnonvariant.filtered3.vcf.gz
bcftools view -S rost -O z -o ROS_NoRelate_NU_Snps.Autosomes.withnonvariant.filtered3.vcf.gz SPI_BJO_ROS_NoRelate_NU_Snps.Autosomes.withnonvariant.filtered3.vcf.gz
bcftools view -S spitsbergen -O z -o SPI_NoRelate_NU_Snps.Autosomes.withnonvariant.filtered3.vcf.gz SPI_BJO_ROS_NoRelate_NU_Snps.Autosomes.withnonvariant.filtered3.vcf.gz

for i in SPI BJO ROS ;
do echo $i
sbatch TajimaSlide.sh $i
done

```

TajimaSlide.sh

```bash


module load Miniconda3/4.9.2
export PS1=\$
source ${EBROOTMINICONDA3}/etc/profile.d/conda.sh
conda deactivate &>/dev/null

conda activate miniconda3_envs/vcf-kit

gunzip ${1}_NoRelate_NU_Snps.Autosomes.withnonvariant.filtered3.vcf.gz ;
vk tajima 50,000 25,000 ${1}_NoRelate_NU_Snps.Autosomes.withnonvariant.filtered3.vcf > ${1}_Tajima_sliding  ;
gzip ${1}_NoRelate_NU_Snps.Autosomes.withnonvariant.filtered3.vcf ;

conda deactivate

```

pi and dxy - per Chromo

```bash
#Do pi + dx per chromo

module purge
module load BCFtools/1.9-intel-2018b
for i in {01..24};
do bcftools view SPI_BJO_ROS_NoRelate_NU_Snps.Autosomes.withnonvariant.filtered3.vcf.gz Chr_${i} | bgzip > SPI_BJO_ROS_NoRelate_NU_Snps.Autosomes.withnonvariant.filtered3.Chromosome_${i}.gz ;
bcftools index SPI_BJO_ROS_NoRelate_NU_Snps.Autosomes.withnonvariant.filtered3.Chromosome_${i}.gz
done

for i in {01..24};
do sbatch VCF_Stats.sh ${i} ;
done

```

VCF_Stats.sh

```bash


module load Miniconda3/4.9.2
export PS1=\$
source ${EBROOTMINICONDA3}/etc/profile.d/conda.sh
conda deactivate &>/dev/null

conda activate miniconda3_envs/simonmartin

python3 parseVCFs.py -i SPI_BJO_ROS_NoRelate_NU_Snps.Autosomes.withnonvariant.filtered3.Chromosome_${1}.gz -o SPI_BJO_ROS_NoRelate_NU_Snps.Autosomes.withnonvariant.filtered3.Chromosome_${1}.geno -t 10 --include Chr_${1}

#Essentially this is a genotype matrix for each of our individuals. It is also a good idea to compress this file with gzip. This is also really straightforward.
gzip SPI_BJO_ROS_NoRelate_NU_Snps.Autosomes.withnonvariant.filtered3.Chromosome_${1}.geno

python3 popgenWindows.py -g SPI_BJO_ROS_NoRelate_NU_Snps.Autosomes.withnonvariant.filtered3.Chromosome_${1}.geno.gz -o Chromosome_${1}_div_stats.csv \
-f phased -w 50000 -m 10 -s 25000 \
-T 4 \
-p SPI -p ROS -p BJO \
--popsFile pop_file \
--writeFailedWindow

conda deactivate

```
