# Add Outgroup

Call SNPs single sample

```bash

mkdir Data
mkdir Log

for i in {02..25} Z Unplaced ;
do echo ${i} ;
sbatch nucl_SNPs_pt1a.sh RAZ ${i} ; 
done ;
echo "W" ;
sbatch nucl_SNPs_pt1b.sh ;

# need it differently for chr1 
# need to split it up into 4 intervals

nano interval.list
Chr_01:1-53877620
Chr_01:53877621-107755240
Chr_01:107755241-161632860
Chr_01:161632861-215510477

cat interval.list | while read line ;
do echo ${line} ;
sbatch nucl_SNPs_pt1a_Chr01.sh RAZ ${line} ; 
done ;

```

nucl_SNPs_pt1a

```bash

module load GATK/4.2.0.0-GCCcore-9.3.0-Java-11

gatk --java-options "-Xmx8g" HaplotypeCaller \
-VS STRICT \
-R Puffin_ReferenceGenome.NU.MT.fasta \
-I RAZ_Anc.Puffin_ReferenceGenome.NU.MT.realigned.bam \
-ERC GVCF \
-L Chr_${2} \
-O RAZ.Chr_${2}.g.vcf.gz 2> Haplotype_caller.RAZ.Chr_${2}.out

# nucl_SNPs_pt1a_Chr01.sh has 
# -L ${2}

mv Haplotype_caller.RAZ.Chr_${2}.out Log/ ;
mv RAZ.Chr_${2}.g.vcf.gz Data/ ;
mv RAZ.Chr_${2}.g.vcf.gz.tbi Data/ ;
```

nucl_SNPs_pt1b

```bash

module load GATK/4.2.0.0-GCCcore-9.3.0-Java-11

gatk --java-options "-Xmx8g" HaplotypeCaller \
-VS STRICT \
-R Puffin_ReferenceGenome.NU.MT.fasta \
-I RAZ_Ancestral/RAZ_Anc.Puffin_ReferenceGenome.NU.MT.realigned.bam \
-ploidy 1 \
-ERC GVCF \
-L Chr_W \
-O RAZ.Chr_W.g.vcf.gz 2> Haplotype_caller.RAZ.Chr_W.out

mv Haplotype_caller.RAZ.Chr_W.out Log/ ;
mv RAZ.Chr_W.g.vcf.gz Data/ ;
mv RAZ.Chr_W.g.vcf.gz.tbi Data/ ;

```

```bash

mkdir VCF

for i in {02..25} Z Unplaced W ; 
do echo ${i} ;
sbatch nucl_SNPs_pt2a.sh ${i} ; 
done

cat ../interval.list | while read line ;
do echo ${line} ;
sbatch nucl_SNPs_Chr01_pt2a.sh ${line} ; 
done ;

# nucl_SNPs_Chr01_pt2a.sh has
# -L ${1}

```

Combine GVCFs and Call Genotypes (GATK4) - per chromosome

nano nucl_SNPs_pt2a.sh

```bash

module load GATK/4.2.0.0-GCCcore-9.3.0-Java-11

mkdir tmp_${1}

gatk --java-options "-Xmx8g" GenotypeGVCFs \
-R Puffin_ReferenceGenome.NU.MT.fasta \
-L Chr_${1} \
-V RAZ.Chr_${1}.g.vcf.gz \
-O RAZ.Chr_${1}.raw.vcf.gz \
--include-non-variant-sites \
--tmp-dir tmp_${1}

rm -r tmp_${1}

```

Combine VCFâ€™s of various chromosomes

```bash

mv RAZ.Chr_*.raw.vcf.gz VCF
mv RAZ.Chr_*.raw.vcf.gz.tbi VCF

cd VCF

module load BCFtools/1.9-intel-2018b

ls | grep "raw.vcf.gz" | grep -v "Chr_W" | grep -v "Chr_Z" | grep -v "Chr_Unplaced" | grep -v ".tbi" > vcf.list
#ls | grep "raw.vcf.gz" | grep -v "Chr_W" | grep -v ".tbi" > vcf_All.list

cat vcf.list | while read line ;
do echo ${line} ;
bcftools index ${line} ;
done ;
bcftools index RAZ.Chr_W.raw.vcf.gz
bcftools index RAZ.Chr_Z.raw.vcf.gz
bcftools index RAZ.Chr_Unplaced.raw.vcf.gz

bcftools concat -n -O z -o RAZ_NU_SnpsAndIndels.Autosomes.raw.vcf.gz -f vcf.list
#-n naive-> faster
#-O output format gzipped vcf
mkdir temp
bcftools sort -o RAZ_NU_SnpsAndIndels.Autosomes.sorted.vcf.gz -O z -T temp RAZ_NU_SnpsAndIndels.Autosomes.raw.vcf.gz
bcftools index RAZ_NU_SnpsAndIndels.Autosomes.sorted.vcf.gz
zcat RAZ_NU_SnpsAndIndels.Autosomes.sorted.vcf.gz | grep -v "#" | wc -l

```

Merge SNPs of RAZ with SNPs of Puffins, while only keeping positions in Puffin SNP vcf

with non-pruned

```bash

cp SPI_BJO_ROS_NoRelate_NU_Snps.Autosomes.filtered4.vcf.gz .

module purge
module load VCFtools/0.1.16-intel-2018b-Perl-5.28.0
module load BCFtools/1.9-intel-2018b

gunzip SPI_BJO_ROS_NoRelate_NU_Snps.Autosomes.filtered4.vcf.gz
bgzip SPI_BJO_ROS_NoRelate_NU_Snps.Autosomes.filtered4.vcf
mkdir temp
bcftools sort -o SPI_BJO_ROS_NoRelate_NU_Snps.Autosomes.filtered4.sorted.vcf.gz -O z -T temp SPI_BJO_ROS_NoRelate_NU_Snps.Autosomes.filtered4.vcf.gz
bcftools index SPI_BJO_ROS_NoRelate_NU_Snps.Autosomes.filtered4.sorted.vcf.gz
zcat SPI_BJO_ROS_NoRelate_NU_Snps.Autosomes.filtered4.sorted.vcf.gz | grep -v "#" | wc -l

mkdir RAZ_Puff_Variant
bcftools isec -p RAZ_Puff_Variant -c all SPI_BJO_ROS_NoRelate_NU_Snps.Autosomes.filtered4.sorted.vcf.gz RAZ_NU_SnpsAndIndels.Autosomes.sorted.vcf.gz
cd RAZ_Puff_Variant
ls *.vcf | while read line ;
do echo $line ;
bgzip ${line} ;
done

mkdir tmp
bcftools sort -o 0002.sorted.vcf.gz -O z -T tmp 0002.vcf.gz 
bcftools index 0002.sorted.vcf.gz
zcat 0002.sorted.vcf.gz | grep -v "#" | wc -l

mkdir tmp
bcftools sort -o 0003.sorted.vcf.gz -O z -T tmp 0003.vcf.gz
bcftools index 0003.sorted.vcf.gz
zcat 0003.sorted.vcf.gz | grep -v "#" | wc -l

bcftools merge -O z -o SPI_BJO_ROS_NoRelate_NU_Snps.Autosomes.withOutgroup.vcf.gz --merge all 0002.sorted.vcf.gz 0003.sorted.vcf.gz
zcat SPI_BJO_ROS_NoRelate_NU_Snps.Autosomes.withOutgroup.vcf.gz | grep -v "#" | wc -l

#biallelic
bcftools view -m2 -M2 -v snps -O z -o SPI_BJO_ROS_NoRelate_NU_Snps.Autosomes.withOutgroup.edit.vcf.gz SPI_BJO_ROS_NoRelate_NU_Snps.Autosomes.withOutgroup.vcf.gz
zcat SPI_BJO_ROS_NoRelate_NU_Snps.Autosomes.withOutgroup.edit.vcf.gz | grep -v "#" | wc -l

#non-missing
bcftools filter -i 'AN==36' -O z -o SPI_BJO_ROS_NoRelate_NU_Snps.Autosomes.withOutgroup.edit2.vcf.gz SPI_BJO_ROS_NoRelate_NU_Snps.Autosomes.withOutgroup.edit.vcf.gz
zcat SPI_BJO_ROS_NoRelate_NU_Snps.Autosomes.withOutgroup.edit2.vcf.gz | grep -v "#" | wc -l

```

with nonvariant (psmc mut rate)

```bash
#get files from Stairway prep
cp SPI_BJO_ROS_NU_Snps.Chr_*.withnonvariant.filtered3.sorted.vcf.gz .

#ALSO
cd /cluster/work/users/oliverke/HighCov/NucAnalysis/GATK/Data/Outgroup/Data/VCF
ls | grep "raw.vcf.gz" | grep "Chr_01:" | grep -v "csi" | grep -v "tbi" > vcf.edit.list
module load BCFtools/1.9-intel-2018b
bcftools concat -n -O z -o RAZ.Chr_01.raw.vcf.gz -f vcf.edit.list

# NEED TO DO IT PER CHROMO
cd /cluster/work/users/oliverke/HighCov/NucAnalysis/GATK/Data/Outgroup

for i in {01..24} ;
do echo $i ;
sbatch NonVariantOutgroup.sh ${i} ;
done ;

```

NonVariantOutgroup.sh

```bash


module purge
module load BCFtools/1.9-intel-2018b

gunzip SPI_BJO_ROS_NU_Snps.Chr_${1}.withnonvariant.filtered3.sorted.vcf.gz
bgzip SPI_BJO_ROS_NU_Snps.Chr_${1}.withnonvariant.filtered3.sorted.vcf
mkdir temp_${1}
bcftools sort -o SPI_BJO_ROS_NU_Snps.Chr_${1}.withnonvariant.sorted2.vcf.gz -O z -T temp_${1} SPI_BJO_ROS_NU_Snps.Chr_${1}.withnonvariant.filtered3.sorted.vcf.gz
bcftools index SPI_BJO_ROS_NU_Snps.Chr_${1}.withnonvariant.sorted2.vcf.gz
zcat SPI_BJO_ROS_NU_Snps.Chr_${1}.withnonvariant.sorted2.vcf.gz | grep -v "#" | wc -l

cp RAZ.Chr_${1}.raw.vcf.gz .
gunzip RAZ.Chr_${1}.raw.vcf.gz
bgzip RAZ.Chr_${1}.raw.vcf
mkdir temp_${1}
bcftools sort -o RAZ.Chr_${1}.raw.sorted2.vcf.gz -O z -T temp_${1} RAZ.Chr_${1}.raw.vcf.gz
bcftools index RAZ.Chr_${1}.raw.sorted2.vcf.gz
zcat RAZ.Chr_${1}.raw.sorted2.vcf.gz | grep -v "#" | wc -l

mkdir RAZ_Puff_NonVariant_${1}
bcftools isec -p RAZ_Puff_NonVariant_${1} -c all SPI_BJO_ROS_NU_Snps.Chr_${1}.withnonvariant.sorted2.vcf.gz RAZ.Chr_${1}.raw.sorted2.vcf.gz
cd RAZ_Puff_NonVariant_${1}
ls *.vcf | while read line ;
do echo $line ;
bgzip ${line} ;
done

mkdir tmp
bcftools sort -o 0002.sorted.vcf.gz -O z -T tmp 0002.vcf.gz 
bcftools index 0002.sorted.vcf.gz
zcat 0002.sorted.vcf.gz | grep -v "#" | wc -l

mkdir tmp
bcftools sort -o 0003.sorted.vcf.gz -O z -T tmp 0003.vcf.gz
bcftools index 0003.sorted.vcf.gz
zcat 0003.sorted.vcf.gz | grep -v "#" | wc -l

bcftools merge -O z -o SPI_BJO_ROS_NU_Snps.Chr_${1}.withOutgroup.nonvariant.vcf.gz --merge all 0002.sorted.vcf.gz 0003.sorted.vcf.gz
zcat SPI_BJO_ROS_NU_Snps.Chr_${1}.withOutgroup.nonvariant.vcf.gz | grep -v "#" | wc -l

#biallelic and mono
bcftools view -M2 -V indels -O z -o SPI_BJO_ROS_NU_Snps.Chr_${1}.withOutgroup.nonvariant.edit.vcf.gz SPI_BJO_ROS_NU_Snps.Chr_${1}.withOutgroup.nonvariant.vcf.gz
zcat SPI_BJO_ROS_NU_Snps.Chr_${1}.withOutgroup.nonvariant.edit.vcf.gz | grep -v "#" | wc -l

#non-missing
bcftools filter -i 'AN==38' -O z -o SPI_BJO_ROS_NU_Snps.Chr_${1}.withOutgroup.nonvariant.edit2.vcf.gz SPI_BJO_ROS_NU_Snps.Chr_${1}.withOutgroup.nonvariant.edit.vcf.gz

```

Then

```bash

# merge the chromosome files
module purge
module load BCFtools/1.9-intel-2018b

touch vcf_All.list
for i in {01..24} ;
do echo -e "RAZ_Puff_NonVariant_${i}/SPI_BJO_ROS_NU_Snps.Chr_${i}.withOutgroup.nonvariant.edit2.vcf.gz" >> vcf_All.list
done

bcftools concat -n -O z -o SPI_BJO_ROS_NU_Snps.Autosomes.withOutgroup.nonvariant.final.vcf.gz -f vcf_All.list
mkdir tmp
bcftools sort -o SPI_BJO_ROS_NU_Snps.Autosomes.withOutgroup.nonvariant.final.sorted.vcf.gz -O z -T tmp SPI_BJO_ROS_NU_Snps.Autosomes.withOutgroup.nonvariant.final.vcf.gz
bcftools index SPI_BJO_ROS_NU_Snps.Autosomes.withOutgroup.nonvariant.final.sorted.vcf.gz
zcat SPI_BJO_ROS_NU_Snps.Autosomes.withOutgroup.nonvariant.final.sorted.vcf.gz | grep -v "#" | wc -l


```
