#PSMC + Stairway

Need a good estimate of mutation rate

```bash

#pairwise divergence
#according to https://onlinelibrary.wiley.com/doi/epdf/10.1111/1755-0998.13457 Prasad et al. 2021
#take mean divergence to RAZ
#use 34 MYA (19-55MYA) as proxy
#http://www.timetree.org/
#mutation rate = pairwise distance X 2 / divergence time
#calculate generational mutation rate with generation time

#OR
# according to https://www.science.org/doi/10.1126/sciadv.abl6496
# take the # of polymorpic sites vs total sites (RAZ vs ref) or (RAZ vs. single sample?)
#use 34 MYA (19-55MYA) as proxy
#http://www.timetree.org/
#calculate generational mutation rate with generation time

#generation time:
#https://conbio.onlinelibrary.wiley.com/doi/full/10.1111/cobi.13486
# 14.2 in SuppTable4

#Map RAZ to new reference
cd /cluster/work/users/oliverke/HighCov/NucAnalysis/PSMC

cp ../SV/samples .

mkdir MutRate
cd MutRate

cat ../samples | while read line ;
do echo ${line} ;
cp ${line}.Puffin_ReferenceGenome.NU.MT.realigned.bam .
cp ${line}.Puffin_ReferenceGenome.NU.MT.realigned.bai .
done

cp RAZ_Anc.Puffin_ReferenceGenome.NU.MT.realigned.bam .
cp RAZ_Anc.Puffin_ReferenceGenome.NU.MT.realigned.bai .

```

IBS/SeqDivergence method

```bash


grep -v "Chr_Z:" ChromosomeList | grep -v "Chr_Unplaced:" | grep -v "Chr_W:" ChromosomeList | grep -v "Chr_25:"  > ChromosomeList.edit

mkdir test
cd test
ls ../*.Puffin_ReferenceGenome.NU.MT.realigned.bam > bams_good.edit
cp bams_good.edit  ../
cd ../
rm -r test

mkdir IBS

sbatch Outgroup_ANGSD_IBS.sh ;

```

Outgroup_ANGSD_IBS.sh

```bash

mkdir temp

cp -r Puffin_ReferenceGenome.NU.MT.fasta* temp/ #reference genome
cp ChromosomeList.edit temp/ #list of chromosomes
cp bams_good.edit temp/   

cd temp/

#same sites as genotype likelihood panel
FILTERS="-uniqueOnly 1 -remove_bads 1 -minMapQ 30 -minQ 30 -C 50 -baq 2 -checkBamHeaders 1 -minInd 19 -setMinDepthInd 3 -setMaxDepthInd 80"

# T O   D O :
# -GL 1 : samtools likelihood model
# -doGlf 2 : output beagle format (for admixture)
# -doGeno 8 : bgenotype likelihoods format for ngsLD
# -makeMatrix 1 -doIBS 1 -doCov 1 : identity-by-state and covariance matrices based on single-read resampling (robust to variation in coverage across samples)
TODO="-doCounts 1 -doIBS 2 -makeMatrix 1"

# Run command:
module load angsd/0.931-GCC-8.2.0-2.31.1

sleep 120 

angsd -b bams_good.edit -GL 1 $FILTERS $TODO -ref Puffin_ReferenceGenome.NU.MT.fasta -rf ChromosomeList.edit -out OutgroupAngsd_IBS

cp -r OutgroupAngsd_IBS* PSMC/MutRate/IBS

cd ../

rm -r temp/

```

```bash

#take mean divergence to RAZ
#RAZ is in column 7 of ibsMat (7th in bams_good)

cd /cluster/work/users/oliverke/HighCov/NucAnalysis/PSMC/MutRate/IBS
DivRate=$(cut -f7 OutgroupAngsd_IBS.ibsMat | grep -v -w "0.000000" | awk '{ total += $1 } END { print total/NR }' ) 
MutRate=$(bc <<<"scale=12; 2 * ${DivRate} / 34000000 ")
#use 34 MYA (19-55MYA) as proxy

GenMutRate=$(bc <<<"scale=12; ${MutRate} * 14.2" )
echo -e "IBS\t${GenMutRate}" >> Comp_IBS_Polymorph.GenMutRate
#calculate generational mutation rate with 14.2 yr generation time

```

Polymorph Sites method

```bash

module purge
module load BCFtools/1.14-GCC-11.2.0

cp SPI_BJO_ROS_NU_Snps.Autosomes.withOutgroup.nonvariant.final.sorted.vcf.gz .

bcftools index SPI_BJO_ROS_NU_Snps.Autosomes.withOutgroup.nonvariant.final.sorted.vcf.gz

for i in {01..24} ;
do echo $i ;
bcftools view -r Chr_${i} -O z -o SPI_BJO_ROS_NU_Snps.Chr_${i}.withOutgroup.nonvariant.final.sorted.vcf.gz SPI_BJO_ROS_NU_Snps.Autosomes.withOutgroup.nonvariant.final.sorted.vcf.gz
mkdir tmp
bcftools sort -o SPI_BJO_ROS_NU_Snps.Chr_${i}.withOutgroup.nonvariant.final.sorted2.vcf.gz -O z -T tmp SPI_BJO_ROS_NU_Snps.Chr_${i}.withOutgroup.nonvariant.final.sorted.vcf.gz
bcftools index SPI_BJO_ROS_NU_Snps.Chr_${i}.withOutgroup.nonvariant.final.sorted2.vcf.gz
done

#Vcffixup
for i in {01..24} ;
do echo $i ;
sbatch VCFfixup.sh ${i} ;
done ;

```

```bash


module load Miniconda3/4.9.2

# Set the ${PS1} (needed in the source of the Anaconda environment)
export PS1=\$

# Source the conda environment setup
# The variable ${EBROOTANACONDA3} or ${EBROOTMINICONDA3}
# So use one of the following lines
# comes with the module load command
# source ${EBROOTANACONDA3}/etc/profile.d/conda.sh
source ${EBROOTMINICONDA3}/etc/profile.d/conda.sh

# Deactivate any spill-over environment from the login node
conda deactivate &>/dev/null

# Activate the environment by using the full path (not name)
# to the environment. The full path is listed if you do
# conda info --envs at the command prompt.

conda activate miniconda3_envs/vcflib

gunzip SPI_BJO_ROS_NU_Snps.Chr_${1}.withOutgroup.nonvariant.final.sorted2.vcf.gz
vcffixup SPI_BJO_ROS_NU_Snps.Chr_${1}.withOutgroup.nonvariant.final.sorted2.vcf | gzip - > SPI_BJO_ROS_NU_Snps.Chr_${1}.withOutgroup.nonvariant.final.sorted2.edit.vcf.gz
gzip SPI_BJO_ROS_NU_Snps.Chr_${1}.withOutgroup.nonvariant.final.sorted2.vcf

conda deactivate

```

```bash

module purge
module load BCFtools/1.14-GCC-11.2.0

for i in {01..24} ;
do echo $i ;
gunzip SPI_BJO_ROS_NU_Snps.Chr_${i}.withOutgroup.nonvariant.final.sorted2.edit.vcf.gz
bgzip SPI_BJO_ROS_NU_Snps.Chr_${i}.withOutgroup.nonvariant.final.sorted2.edit.vcf
bcftools index SPI_BJO_ROS_NU_Snps.Chr_${i}.withOutgroup.nonvariant.final.sorted2.edit.vcf.gz
done ;

ls SPI_BJO_ROS_NU_Snps.Chr_*.withOutgroup.nonvariant.final.sorted2.edit.vcf.gz > vcf.list
bcftools concat -n -O z -o SPI_BJO_ROS_NU_Snps.AllAutosomes.withOutgroup.nonvariant.final.sorted2.edit.vcf.gz -f vcf.list
#-n naive-> faster
#-O output format gzipped vcf
mkdir temp
bcftools sort -o SPI_BJO_ROS_NU_Snps.AllAutosomes.withOutgroup.nonvariant.final.sorted3.edit.vcf.gz -O z -T temp SPI_BJO_ROS_NU_Snps.AllAutosomes.withOutgroup.nonvariant.final.sorted2.edit.vcf.gz
bcftools index SPI_BJO_ROS_NU_Snps.AllAutosomes.withOutgroup.nonvariant.final.sorted3.edit.vcf.gz

#get sample names
bcftools query -l SPI_BJO_ROS_NU_Snps.AllAutosomes.withOutgroup.nonvariant.final.sorted3.edit.vcf.gz > samples

#for each sample that is not RAZ
touch DivRatePerSample
TotalSites=$(zcat SPI_BJO_ROS_NU_Snps.AllAutosomes.withOutgroup.nonvariant.final.sorted3.edit.vcf.gz | grep -v "#" | wc -l | sed 's/ //g' )
echo $TotalSites 
cat samples | grep -v "RAZ_Anc" | while read sample ;
do echo ${sample} ;
bcftools view -s ${sample},RAZ_Anc -O z -o ${sample}_RAZ_Snps.Autosomes.withOutgroup.nonvariant.final.sorted.vcf.gz SPI_BJO_ROS_NU_Snps.AllAutosomes.withOutgroup.nonvariant.final.sorted3.edit.vcf.gz
PolySites=$(bcftools filter -i 'AC==1 || AC==2 || AC==3' ${sample}_RAZ_Snps.Autosomes.withOutgroup.nonvariant.final.sorted.vcf.gz | grep -v "#" | wc -l | sed 's/ //g' ) ;
DivRate=$(bc <<<"scale=12; ${PolySites} / ${TotalSites} ") ;
echo ${DivRate} >> DivRatePerSample ;
done

MutRateSample=$(cut -f1 DivRatePerSample | awk '{ total += $1 } END { print total/NR }' ) 
MutRateYear=$(bc <<<"scale=12; ${MutRateSample} / 34000000 ")
#use 34 MYA (19-55MYA) as proxy
#http://www.timetree.org/
GenMutRate=$(bc <<<"scale=12; ${MutRateYear} * 14.2" )
mv ../IBS/Comp_IBS_Polymorph.GenMutRate ../
echo -e "Poly\t${GenMutRate}" >> ../Comp_IBS_Polymorph.GenMutRate

less ../Comp_IBS_Polymorph.GenMutRate

#Generational mutation rate
cut -f2 Comp_IBS_Polymorph.GenMutRate | awk '{ total += $1 } END { print total/NR }'

```

PSMC

```bash
cd /cluster/work/users/oliverke/HighCov/NucAnalysis/PSMC

#generate coverage table for each sample
#copy from local table
scp CoveragePuffin oliverke@saga.sigma2.no:xxx

#SAGA
#Only Take Autosomes
touch name.lst
for i in {01..24} ;
do echo -e "Chr_${i}" >> name.lst ;
done

#TEST
cat samples | grep "SPI" | head -1 | while read line ;
do echo ${line} ;
Coverage=$(cat CoveragePuffin | grep -w ${line} | cut -f18)
echo ${Coverage} ;
DoubleCov=$(echo $(printf %.$2f $(echo "scale=2;${Coverage}*2" | bc)))
echo ${DoubleCov} ;
sbatch PSMC_test.sh ${line} ${DoubleCov} ;
done

cat samples | grep "ROS" | head -1 | while read line ;
do echo ${line} ;
Coverage=$(cat CoveragePuffin | grep -w ${line} | cut -f18)
echo ${Coverage} ;
DoubleCov=$(echo $(printf %.$2f $(echo "scale=2;${Coverage}*2" | bc)))
echo ${DoubleCov} ;
sbatch PSMC_test.sh ${line} ${DoubleCov} ;
done

cat samples | grep "BJO" | head -1 | while read line ;
do echo ${line} ;
Coverage=$(cat CoveragePuffin | grep -w ${line} | cut -f18)
echo ${Coverage} ;
DoubleCov=$(echo $(printf %.$2f $(echo "scale=2;${Coverage}*2" | bc)))
echo ${DoubleCov} ;
sbatch PSMC_test.sh ${line} ${DoubleCov} ;
done

```

PSMC_test.sh

```bash

module purge
module load psmc/0.6.5-foss-2018b
module load SAMtools/1.9-foss-2018b
module load BCFtools/1.9-foss-2018b

samtools mpileup -C50 -uf Puffin_ReferenceGenome.NU.MT.fasta \
${1}.Puffin_ReferenceGenome.NU.MT.realigned.bam \
| bcftools call -c - \
| vcfutils.pl vcf2fq -d 10 -D ${2} -Q 25 -l 10 | gzip > ${1}.diploid.fq.gz 

# -D mean coverage X 2
# other filters according to Krystyna Nadachowska-Brzyska et al. 2015 Temporal Dynamics of Avian Populations during Pleistocene Revealed by Whole-Genome Sequences
# and default and Krystyna Nadachowska-Brzyska et al. 2016 flycatcher (-d 10)

gunzip ${1}.diploid.fq.gz 

module purge 
module load seqtk/1.3-foss-2018b

seqtk subseq ${1}.diploid.fq name.lst | seqtk seq -l60 - > ${1}.diploid.autosomes.fq
gzip ${1}.diploid.autosomes.fq
#takes 15 min

module purge
module load psmc/0.6.5-foss-2018b
module load SAMtools/1.9-foss-2018b
module load BCFtools/1.9-foss-2018b

fq2psmcfa -q20 ${1}.diploid.autosomes.fq.gz > ${1}.diploid.psmcfa
#1min

splitfa ${1}.diploid.psmcfa > ${1}.diploid.split.psmcfa
#1min

# bird paper Nadachowska-Brzyska et al. 2015
psmc -N30 -t5 -r5 -p "4+30*2+4+6+10" -o ${1}.diploid_test1.psmc ${1}.diploid.psmcfa
#4hrs

# default and Foote
psmc -N25 -t15 -r5 -p "4+25*2+4+6" -o ${1}.diploid_test2.psmc ${1}.diploid.psmcfa
#4hrs

# Rebecca Taylor
psmc -N25 -t5 -r5 -p "4+25*2+4+6" -o ${1}.diploid_test3.psmc ${1}.diploid.psmcfa
#4hrs

#psmc2history.pl ${1}.diploid.psmc | history2ms.pl > ${1}.ms-cmd.sh
#1min

```

Then

```bash

module purge
module load psmc/0.6.5-foss-2018b
module swap GCCcore/7.3.0 GCCcore/8.2.0
module swap zlib/1.2.11-GCCcore-7.3.0 zlib/1.2.11-GCCcore-8.2.0
module load gnuplot/5.2.6-GCCcore-8.2.0

MutRate=$(cut -f2 MutRate/Comp_IBS_Polymorph.GenMutRate | awk '{ total += $1 } END { print total/NR }' ) 
MutRate_Science=$(printf "%.2e\n" ${MutRate})

for line in ROS001 BJO001 SPI001 ;
do echo ${line} ;
psmc_plot.pl -u ${MutRate_Science} -X 0 -x 10000 -Y 0 -g 14.2 -R ${line}.diploid_test1 ${line}.diploid_test1.psmc ;
psmc_plot.pl -u ${MutRate_Science} -X 0 -x 10000 -Y 0 -g 14.2 -R ${line}.diploid_test2 ${line}.diploid_test2.psmc ;
psmc_plot.pl -u ${MutRate_Science} -X 0 -x 10000 -Y 0 -g 14.2 -R ${line}.diploid_test3 ${line}.diploid_test3.psmc ;
done 
#1min

# according to https://www.nature.com/articles/s41437-021-00443-8.pdf?proof=t and Rebecca
# The output.psmc file was evaluated to see if a sufficient number of recombination
# events (5th column of the file) occurred in each atomic interval after 20 iterations.

#DO that for each test ... decide which is "best" - if all ok refer to default

#check plotting differences

touch TestResults.txt
for line in ROS001 BJO001 SPI001 ;
do echo $line ;
echo $line >> TestResults.txt ;
for i in test1 test2 test3 ;
do echo $i ;
echo $i >> TestResults.txt ;
Value=$(cat ${line}.diploid_${i}.psmc | grep "RD\t20" -A6 | tail -1 | awk -F ':' '{print $3}') ;
echo $Value >> TestResults.txt ;
done ;
echo -e "\n" >> TestResults.txt ;
done 

# also cat ${line}.diploid_${i}.psmc | grep "RD\t20" -A100
# it turns out that (as Rebecca says in the Caribou paper) the p-parameter doesn't really
# change anything.
# however, t=15 doesn't reach 10 rec. in each interval after 20 rounds.
# -> use test1 or test3 -> test1 as it is bird specific

#THEN

mkdir Test
mv *test* Test/
rm slurm*
rm *diploid*

cat samples | while read line ;
do echo ${line} ;
Coverage=$(cat CoveragePuffin | grep -w ${line} | cut -f18)
echo ${Coverage} ;
DoubleCov=$(echo $(printf %.$2f $(echo "scale=2;${Coverage}*2" | bc)))
echo ${DoubleCov} ;
sbatch PSMC_normal.sh ${line} ${DoubleCov} ;
done

cat samples | while read line ;
do echo ${line} ;
sbatch PSMC_normal.sh ${line} ;
done 
```

PSMC_normal.sh

```bash


module purge
module load psmc/0.6.5-foss-2018b
module load SAMtools/1.9-foss-2018b
module load BCFtools/1.9-foss-2018b

samtools mpileup -C50 -uf Puffin_ReferenceGenome.NU.MT.fasta \
${1}.Puffin_ReferenceGenome.NU.MT.realigned.bam \
| bcftools call -c - \
| vcfutils.pl vcf2fq -d 10 -D ${2} -Q 25 -l 10 | gzip > ${1}.diploid.fq.gz 
#takes 12 hrs, 2G
# -D mean coverage X 2
# other filters according to Krystyna Nadachowska-Brzyska et al. 2015 Temporal Dynamics of Avian Populations during Pleistocene Revealed by Whole-Genome Sequences
# and default and Krystyna Nadachowska-Brzyska et al. 2016 flycatcher (-d 10)

gunzip ${1}.diploid.fq.gz 

module purge 
module load seqtk/1.3-foss-2018b

seqtk subseq ${1}.diploid.fq name.lst | seqtk seq -l60 - > ${1}.diploid.autosomes.fq
gzip ${1}.diploid.autosomes.fq
#takes 15 min

module purge
module load psmc/0.6.5-foss-2018b
module load SAMtools/1.9-foss-2018b
module load BCFtools/1.9-foss-2018b

fq2psmcfa -q20 ${1}.diploid.autosomes.fq.gz > ${1}.diploid.psmcfa
#1min

splitfa ${1}.diploid.psmcfa > ${1}.diploid.split.psmcfa
#1min

psmc -N30 -t5 -r5 -p "4+30*2+4+6+10" -o ${1}.diploid.psmc ${1}.diploid.psmcfa
#4hrs

#psmc2history.pl ${1}.diploid.psmc | history2ms.pl > ${1}.ms-cmd.sh
#1min
```

Then

```bash


cat samples | while read line ;
do echo ${line} ;
for i in {001..100} ;
do echo ${i} ;
sbatch PSMC_bootstrap.sh ${line} ${i} ;
done ;
done ;
```

PSMC_bootstrap.sh

```bash

module purge
module load psmc/0.6.5-foss-2018b
module load SAMtools/1.9-foss-2018b
module load BCFtools/1.9-foss-2018b

mkdir ${1}.tmp.${2}

cd ${1}.tmp.${2}

cp ../${1}.diploid.split.psmcfa .

psmc -N30 -t5 -r5 -b -p "4+30*2+4+6+10" -o ${1}.round-${2}.psmc ${1}.diploid.split.psmcfa

cp ${1}.round-${2}.psmc ../

cd ..

rm -r ${1}.tmp.${2}
```

Plotting

```bash

rm slurm*

module purge
module load psmc/0.6.5-foss-2018b
module swap GCCcore/7.3.0 GCCcore/8.2.0
module swap zlib/1.2.11-GCCcore-7.3.0 zlib/1.2.11-GCCcore-8.2.0
module load gnuplot/5.2.6-GCCcore-8.2.0

cat samples | while read line ;
do echo ${line} ;
cat ${line}.diploid.psmc ${line}.round-*.psmc > ${line}.bootstrapped.psmc ;
rm ${line}.round-*.psmc ;
done

MutRate_Science=1.7125e-08

cat samples| while read line ;
do echo ${line} ;
psmc_plot.pl -u ${MutRate_Science} -X 0 -x 0 -Y 0 -g 14.2 -R ${line}.bootstrapped ${line}.bootstrapped.psmc ;
rm ${line}.bootstrapped.replicates.txt
touch ${line}.bootstrapped.replicates.txt
for i in {1..100} ;
do echo ${i} ;
cut -f2 ${line}.bootstrapped.${i}.txt | paste ${line}.bootstrapped.replicates.txt - > tmp ;
mv tmp ${line}.bootstrapped.replicates.txt ;
rm ${line}.bootstrapped.${i}.txt
done
done


#To read into R
touch AllCombined
cat samples | while read line ;
do echo $line ;
cut -f1,2 ${line}.bootstrapped.0.txt | paste - ${line}.bootstrapped.replicates.txt | awk -F '\t' -v OFS='\t' -v id=${line} '{ $(NF+1) = id; print }' > ${line}.combined.txt
cat ${line}.combined.txt >> AllCombined ;
done

```

Stairway

Make SFS first

SFS

```bash

mkdir Stairway
cd Stairway

mkdir VCF
cd VCF

#Per pop and then per chromosome

cp SPI_BJO_ROS_NU_SnpsAndIndels.*.raw.withnonvariant.vcf.gz* .

#Total Variants
zcat SPI_BJO_ROS_NU_SnpsAndIndels.Autosomes.raw.withnonvariant.vcf.gz | grep -v "#" | wc -l

module purge
module load VCFtools/0.1.16-intel-2018b-Perl-5.28.0
module load BCFtools/1.9-intel-2018b

for i in {01..24} ;
do echo ${i} ;
bcftools view -r Chr_${i} -Oz -o SPI_BJO_ROS_NU_SnpsAndIndels.Chr_${i}.raw.withnonvariant.vcf.gz SPI_BJO_ROS_NU_SnpsAndIndels.Autosomes.raw.withnonvariant.vcf.gz ;
done 

for i in {01..24} ;
do echo ${i} ;
sbatch FilterNonVariant.sh ${i} ;
done 

```

FilterNonVariant.sh

```bash

module purge
module load VCFtools/0.1.16-intel-2018b-Perl-5.28.0
module load BCFtools/1.9-intel-2018b

mkdir tmp.${1}
cd tmp.${1}

cp ../SPI_BJO_ROS_NU_SnpsAndIndels.Chr_${1}.raw.withnonvariant.vcf.gz .

#Total Variants
zcat SPI_BJO_ROS_NU_SnpsAndIndels.Chr_${1}.raw.withnonvariant.vcf.gz | grep -v "#" | wc -l > Chr_${1}.count1

#Only SNPs 
module purge
module load VCFtools/0.1.16-intel-2018b-Perl-5.28.0
module load BCFtools/1.9-intel-2018b
vcftools --gzvcf SPI_BJO_ROS_NU_SnpsAndIndels.Chr_${1}.raw.withnonvariant.vcf.gz --remove-indels --out SPI_BJO_ROS_NU_Snps.Chr_${1}.raw.withnonvariant

#Apply GATK filters
gunzip SPI_BJO_ROS_NU_SnpsAndIndels.Chr_${1}.raw.withnonvariant.vcf.gz
bgzip SPI_BJO_ROS_NU_SnpsAndIndels.Chr_${1}.raw.withnonvariant.vcf
bcftools index SPI_BJO_ROS_NU_SnpsAndIndels.Chr_${1}.raw.withnonvariant.vcf.gz
bcftools filter --SnpGap 10 -e 'QD < 2.0 || MQ < 40 || FS > 60.0 || SOR > 3 || MQRankSum < -12.5 || MQRankSum > 12.5 || ReadPosRankSum > 8.0 || ReadPosRankSum < -8.0' -O z -o SPI_BJO_ROS_NU_SnpsAndIndels.Chr_${1}.withnonvariant.filtered1.vcf.gz SPI_BJO_ROS_NU_SnpsAndIndels.Chr_${1}.raw.withnonvariant.vcf.gz
bcftools view -H SPI_BJO_ROS_NU_SnpsAndIndels.Chr_${1}.withnonvariant.filtered1.vcf.gz | wc -l > Chr_${1}.count2
bcftools view -v snps -H SPI_BJO_ROS_NU_SnpsAndIndels.Chr_${1}.withnonvariant.filtered1.vcf.gz | wc -l >> Chr_${1}.count2

ml purge
ml Miniconda3/4.9.2
export PS1=\$
source ${EBROOTMINICONDA3}/etc/profile.d/conda.sh
conda deactivate &>/dev/null
conda activate miniconda3_envs/vcflib
gunzip SPI_BJO_ROS_NU_SnpsAndIndels.Chr_${1}.withnonvariant.filtered1.vcf.gz
vcffixup SPI_BJO_ROS_NU_SnpsAndIndels.Chr_${1}.withnonvariant.filtered1.vcf | gzip - > SPI_BJO_ROS_NU_SnpsAndIndels.Chr_${1}.withnonvariant.filtered1.edit.vcf.gz
gzip SPI_BJO_ROS_NU_SnpsAndIndels.Chr_${1}.withnonvariant.filtered1.vcf
conda deactivate

module purge
module load VCFtools/0.1.16-intel-2018b-Perl-5.28.0
module load BCFtools/1.9-intel-2018b

gunzip SPI_BJO_ROS_NU_SnpsAndIndels.Chr_${1}.withnonvariant.filtered1.edit.vcf.gz
bgzip SPI_BJO_ROS_NU_SnpsAndIndels.Chr_${1}.withnonvariant.filtered1.edit.vcf
bcftools index SPI_BJO_ROS_NU_SnpsAndIndels.Chr_${1}.withnonvariant.filtered1.edit.vcf.gz
bcftools view -H SPI_BJO_ROS_NU_SnpsAndIndels.Chr_${1}.withnonvariant.filtered1.edit.vcf.gz | wc -l > Chr_${1}.count3
bcftools view -v snps -H SPI_BJO_ROS_NU_SnpsAndIndels.Chr_${1}.withnonvariant.filtered1.edit.vcf.gz | wc -l >> Chr_${1}.count3

# Exclude Indels and only include biallelic
vcftools --gzvcf SPI_BJO_ROS_NU_SnpsAndIndels.Chr_${1}.withnonvariant.filtered1.edit.vcf.gz --remove-indels --max-alleles 2 --recode --recode-INFO-all --stdout | gzip > SPI_BJO_ROS_NU_Snps.Chr_${1}.withnonvariant.vcf.gz
zcat SPI_BJO_ROS_NU_Snps.Chr_${1}.withnonvariant.vcf.gz | grep -v "#" | wc -l > Chr_${1}.count4

#bcftools filter -e 'AC==0 || AC==AN' SPI_BJO_ROS_NU_SnpsAndIndels.filtered2.vcf.gz | bcftools view -m2 -M2 -v snps -O z -o SPI_BJO_ROS_NU_Snps.filtered.vcf.gz
#bcftools view -H SPI_BJO_ROS_NU_Snps.filtered.vcf.gz | wc -l

rm SPI_BJO_ROS_NU_SnpsAndIndels.Chr_${1}.raw.withnonvariant.vcf.gz
cp * ../
cd ../
rm -r tmp.${1}
```

Filter more

```bash
mkdir tmp
cd tmp
mv ../slurm* .
mv ../*count* .
mv ../SPI_BJO_ROS_NU_SnpsAndIndels.Autosomes.raw.withnonvariant.vcf.gz .
mv ../SPI_BJO_ROS_NU_SnpsAndIndels.Autosomes.raw.withnonvariant.vcf.gz.csi .
mv ../*.sh .

for i in {01..24} ;
do echo ${i} ;
mv ../SPI_BJO_ROS_NU_SnpsAndIndels.Chr_${i}.raw.withnonvariant.vcf.gz .
mv ../SPI_BJO_ROS_NU_Snps.Chr_${i}.withnonvariant.vcf.gz .
done

cd ../
rm *.gz*
mv tmp/* .

rm -r tmp/

for i in {01..24} ;
do echo ${i} ;
sbatch FilterNonVariant2.sh ${i} ;
done 

```

FilterNonVariant2.sh

```bash

module purge
module load VCFtools/0.1.16-intel-2018b-Perl-5.28.0
module load BCFtools/1.9-intel-2018b

mkdir tmp.${1}
cd tmp.${1}

#GLOBAL DEPTH
#maxDP 500
#minDP 250
cp ../SPI_BJO_ROS_NU_Snps.Chr_${1}.withnonvariant.vcf.gz .
gunzip SPI_BJO_ROS_NU_Snps.Chr_${1}.withnonvariant.vcf.gz
bgzip SPI_BJO_ROS_NU_Snps.Chr_${1}.withnonvariant.vcf
bcftools index SPI_BJO_ROS_NU_Snps.Chr_${1}.withnonvariant.vcf.gz
bcftools filter -e 'INFO/DP[*] > 500 | INFO/DP[*] < 250 ' -O z -o SPI_BJO_ROS_NU_Snps.Chr_${1}.withnonvariant.filtered1.vcf.gz SPI_BJO_ROS_NU_Snps.Chr_${1}.withnonvariant.vcf.gz
zcat SPI_BJO_ROS_NU_Snps.Chr_${1}.withnonvariant.filtered1.vcf.gz | grep -v "#" | wc -l > Chr_${1}.count5
rm SPI_BJO_ROS_NU_Snps.Chr_${1}.withnonvariant.vcf.gz

#OTHER FILTERS
#--max-missing 1
#--max-meanDP 43.2 
#--min-meanDP 10.8
#--minQ 30 - cannot do minQ because invariant have 0

vcftools --gzvcf SPI_BJO_ROS_NU_Snps.Chr_${1}.withnonvariant.filtered1.vcf.gz --max-missing 1 --max-meanDP 43.4 --min-meanDP 10.9 --remove-filtered-all --recode --recode-INFO-all --stdout | gzip > SPI_BJO_ROS_NU_Snps.Chr_${1}.withnonvariant.filtered2.vcf.gz

gunzip SPI_BJO_ROS_NU_Snps.Chr_${1}.withnonvariant.filtered3.vcf.gz
bgzip SPI_BJO_ROS_NU_Snps.Chr_${1}.withnonvariant.filtered3.vcf
bcftools index SPI_BJO_ROS_NU_Snps.Chr_${1}.withnonvariant.filtered3.vcf.gz

cp * ../
cd ../
rm -r tmp.${1}
```

Then

```bash
module purge
module load BCFtools/1.9-intel-2018b

for i in {01..24} ;
do echo ${i} ;
gunzip SPI_BJO_ROS_NU_Snps.Chr_${i}.withnonvariant.filtered3.vcf.gz
bgzip SPI_BJO_ROS_NU_Snps.Chr_${i}.withnonvariant.filtered3.vcf
mkdir tmp_${i}
bcftools sort -Oz -o SPI_BJO_ROS_NU_Snps.Chr_${i}.withnonvariant.filtered3.sorted.vcf.gz -T tmp_${i} SPI_BJO_ROS_NU_Snps.Chr_${i}.withnonvariant.filtered3.vcf.gz
bcftools index SPI_BJO_ROS_NU_Snps.Chr_${i}.withnonvariant.filtered3.sorted.vcf.gz
done

rm SPI_BJO_ROS_NU_Snps.Chr_*.withnonvariant.filtered3.vcf.gz
rm SPI_BJO_ROS_NU_Snps.Chr_*.withnonvariant.filtered3.vcf.gz.csi
rm SPI_BJO_ROS_NU_Snps.Chr_*.withnonvariant.filtered1.vcf.gz
rm SPI_BJO_ROS_NU_Snps.Chr_*.withnonvariant.filtered1.vcf.gz.csi
rm SPI_BJO_ROS_NU_Snps.Chr_*.withnonvariant.filtered2.vcf.gz 
rm SPI_BJO_ROS_NU_Snps.Chr_*.withnonvariant.filtered2.vcf.gz.csi 

ls SPI_BJO_ROS_NU_Snps.Chr_*.withnonvariant.filtered3.sorted.vcf.gz > chromolist

```

Then easySFS

```bash

module purge
module load BCFtools/1.9-intel-2018b

#separate by pop
grep "BJO" samples > bjornoya
grep "ROS" samples > rost
grep "SPI" samples > spitsbergen
echo "SPI002" >> spitsbergen

for i in {01..24} ;
do echo ${i} ;
bcftools view -S bjornoya -O z -o BJO_NU_Snps.Chr_${i}.withnonvariant.filtered3.sorted.vcf.gz SPI_BJO_ROS_NU_Snps.Chr_${i}.withnonvariant.filtered3.sorted.vcf.gz ;
bcftools view -S rost -O z -o ROS_NU_Snps.Chr_${i}.withnonvariant.filtered3.sorted.vcf.gz SPI_BJO_ROS_NU_Snps.Chr_${i}.withnonvariant.filtered3.sorted.vcf.gz ;
bcftools view -S spitsbergen -O z -o SPI_NU_Snps.Chr_${i}.withnonvariant.filtered3.sorted.vcf.gz SPI_BJO_ROS_NU_Snps.Chr_${i}.withnonvariant.filtered3.sorted.vcf.gz ;
done

# make sfs pop file 
touch BJO.pop
cat bjornoya | while read line ;
do echo -e ${line}"\tBjornoya" >> BJO.pop
done

touch ROS.pop
cat rost | while read line ;
do echo -e ${line}"\tRost" >> ROS.pop
done

touch SPI.pop
cat spitsbergen | while read line ;
do echo -e ${line}"\tSpitsbergen" >> SPI.pop
done

#### Run Easy SFS

for pop in SPI BJO ROS ;
do echo ${pop} ;
for i in {01..24} ; 
#for i in {01..13} ; 
#for i in {01..04} ;
do echo ${i}
sbatch easySFS.sh ${pop} ${i} ;
done
done

```

nano easySFS.sh

```bash

mkdir temp_${1}_${2}
cp ${1}.pop temp_${1}_${2}
cp ${1}_NU_Snps.Chr_${2}.withnonvariant.filtered3.sorted.vcf.gz temp_${1}_${2}
cd temp_${1}_${2}

module purge
module load Miniconda3/4.9.2
export PS1=\$
source ${EBROOTMINICONDA3}/etc/profile.d/conda.sh
conda deactivate &>/dev/null
conda activate miniconda3_envs/easySFS

VCF=${1}_NU_Snps.Chr_${2}.withnonvariant.filtered3.sorted.vcf.gz

easySFS.py -i ${VCF} -p ${1}.pop -f -a --proj=12

mv output output_${1}_${2}
mv output_${1}_${2} ../

cd ../
rm -r temp_${1}_${2}

```

```bash

for pop in SPI BJO ROS ;
do echo ${pop} ;
touch tmp ;
colony=$(cat ${pop}.pop | cut -f2 | head -1)
for i in {01..24} ;
do cat output_${pop}_${i}/fastsimcoal2/${colony}_MAFpop0.obs | head -3 | tail -1 >> tmp
done ;
for col in {1..13} ;
do cut -d ' ' -f${col} tmp | awk '{sum+=$1;} END{print sum;}' >> tmp2
done 
cat tmp2 | tr '\n' ' ' > ${pop}.sfs
rm tmp
rm tmp2
done ;

cd ..
mkdir Runs
cd Runs

#You need to adjust 
- L OR num. loci 
- mu OR mutation rate ; MutRate_Science # assumed mutation rate per site per generation, Smeds et al 2016 collared flycatcher 
- generation time ; 14.2 #according to IUCN
- pop id
- nseq OR sample size (as the number of haploid sequences)
- SFS (no monomorphic)
- nrand #4 numbers, roughly (nseq-2)/4, (nseq-2)/2, (nseq-2)*3/4, nseq-2 ; rounded integers
- proj_dir (unique for each pop; POP_stairway)
- plot_title (No space in it -> output files are named after it)

#Blueprints
MutRate_Science=1.7125e-08

for i in BJO SPI ROS ;
do touch ${i}.blueprint ;
cp ${i}.sfs . ;
NumSites=$(awk -F ' ' '{for(i=1;i<=NF;i++)x+=$i;print x}' ${i}.sfs)
echo "#example blueprint file" >> ${i}.blueprint ;
echo "#input setting" >> ${i}.blueprint ;
echo "popid: "${i} >> ${i}.blueprint ;
echo "nseq: 12" >> ${i}.blueprint ;
echo "L: "${NumSites} >> ${i}.blueprint ;
echo "whether_folded: true" >> ${i}.blueprint ;
SFS=$(cat ${i}.sfs | cut -d ' ' -f2-7) ;
echo -e "SFS: "${SFS} >> ${i}.blueprint ;
echo "#smallest_size_of_SFS_bin_used_for_estimation: 2" >> ${i}.blueprint ;
echo "#largest_size_of_SFS_bin_used_for_estimation: 12" >> ${i}.blueprint ;
echo "pct_training: 0.67" >> ${i}.blueprint ;
echo "nrand: 3 5 8 10" >> ${i}.blueprint ;
echo "project_dir: "${i} >> ${i}.blueprint ;
echo "stairway_plot_dir: stairway_plot_es" >> ${i}.blueprint ;
echo "ninput: 200" >> ${i}.blueprint ;
echo "#output setting" >> ${i}.blueprint ;
echo "mu: "${MutRate_Science} >> ${i}.blueprint ;
echo "year_per_generation: 14.2" >> ${i}.blueprint ;
echo "#plot setting" >> ${i}.blueprint ;
echo "plot_title: "${i} >> ${i}.blueprint ;
echo "xrange: 0.1,6000" >> ${i}.blueprint ;
echo "yrange: 0,0" >> ${i}.blueprint ;
echo "xspacing: 2" >> ${i}.blueprint ;
echo "yspacing: 2" >> ${i}.blueprint ;
echo "fontsize: 12" >> ${i}.blueprint ;
done 

#ADJUST TIME FOR SCRIPT
#THEN
module purge
module load Java/1.8.0_212
cp -r stairway_plot_es .
for i in SPI BJO ROS ;
do echo ${i} ;
java -cp stairway_plot_es Stairbuilder ${i}.blueprint ;
head -802 ${i}.blueprint.sh | sed 's/java/srun -n1 -N1 java/g' | sed 's/$/ \&/g' | sed 's/date \&/date/g' | sed '31~30s/\&//' - | sed '802 s/\&//' - | sed '31~30s/$/\nwait/' - > header.${i}.blueprint.parallel ;
echo "wait" >> header.${i}.blueprint.parallel ;
tail -n+803 ${i}.blueprint.sh | cat header.${i}.blueprint.parallel - > ${i}.blueprint.parallel.sh ;
sbatch Stairwayplot.sh ${i} ;
done

```

nano Stairwayplot.sh

```bash


mkdir temp.${1}
cp -r stairway_plot_es temp.${1} 
cp ${1}.blueprint.parallel.sh temp.${1}
cp -r ${1} temp.${1}
cp ${1}.blueprint temp.${1}
cd temp.${1}/

module load Java/1.8.0_212
bash ${1}.blueprint.parallel.sh

rm -r stairway_plot_es 
rm ${1}.blueprint.parallel.sh
rm ${1}.blueprint
cp -r * ../
cd ../
rm -r temp.${1}

```

```bash

#on Local
#Can plot multiple into 1 with coloring, based on .summary file, but need to add a 0 row in the summary for plotting

for i in SPI BJO ROS ;
do echo ${i} ;
LineToAdd=$(head -2 ${i}.final.summary | tail -1 | awk '{print $1"\t"$2"\t"$3"\t"$4"\t"$5"\t0\t"$7"\t"$8"\t"$9"\t"$10"\t"$11}') ;
head -1 ${i}.final.summary > ${i}.final.edit.summary ;
echo ${LineToAdd} >> ${i}.final.edit.summary ;
tail -n+2 ${i}.final.summary >> ${i}.final.edit.summary ;
done

```
