# MSMC2

```bash

#need all files first

mkdir MSMC2
cd MSMC2

#phased dataset
cp SPI_BJO_ROS.NoRelate.Chr_*.phased.vcf.gz* .

#dataset with nonvariant sites
cp SPI_BJO_ROS_NU_Snps.Chr_*.withnonvariant.filtered3.sorted.vcf.gz* .

#samples
cp samples .

#ref genome
cp Puffin_ReferenceGenome.NU.MT.fasta .

#helper scripts
git clone https://github.com/stschiff/msmc-tools.git

#THEN
module purge
module load BCFtools/1.9-intel-2018b


for i in {01..24} ;
do echo ${i} ;
#take out non-variants + related
bcftools filter -e 'AC!=0' SPI_BJO_ROS_NU_Snps.Chr_${i}.withnonvariant.filtered3.sorted.vcf.gz | bcftools view -S samples -O z -o SPI_BJO_ROS_NU_Snps.Chr_${i}.nonvariant.vcf.gz
bcftools index SPI_BJO_ROS_NU_Snps.Chr_${i}.nonvariant.vcf.gz
#merge with phased
zcat SPI_BJO_ROS.NoRelate.Chr_${i}.phased.vcf.gz | grep "##" > Chr_${i}.header1
zcat SPI_BJO_ROS_NU_Snps.Chr_${i}.nonvariant.vcf.gz | grep "#" | grep -v -f Chr_${i}.header1 > Chr_${i}.header2
zcat SPI_BJO_ROS.NoRelate.Chr_${i}.phased.vcf.gz | grep -v "#" > Chr_${i}.tmp1
zcat SPI_BJO_ROS_NU_Snps.Chr_${i}.nonvariant.vcf.gz | grep -v "#" > Chr_${i}.tmp2
cat Chr_${i}.header1 Chr_${i}.header2 Chr_${i}.tmp1 Chr_${i}.tmp2 | gzip - > SPI_BJO_ROS_NU_Snps.Chr_${i}.msmc2.vcf.gz
rm Chr_${i}.header1 Chr_${i}.header2 Chr_${i}.tmp1 Chr_${i}.tmp2
bcftools sort SPI_BJO_ROS_NU_Snps.Chr_${i}.msmc2.vcf.gz > SPI_BJO_ROS_NU_Snps.Chr_${i}.msmc2.sorted.vcf.gz
mv SPI_BJO_ROS_NU_Snps.Chr_${i}.msmc2.sorted.vcf.gz SPI_BJO_ROS_NU_Snps.Chr_${i}.msmc2.sorted.vcf
bgzip SPI_BJO_ROS_NU_Snps.Chr_${i}.msmc2.sorted.vcf
bcftools index SPI_BJO_ROS_NU_Snps.Chr_${i}.msmc2.sorted.vcf.gz
cat samples | while read ind ;
do echo $ind ;
bcftools view -s ${ind} -O z -o ${ind}_NU_Snps.Chr_${i}.msmc2.sorted.vcf.gz SPI_BJO_ROS_NU_Snps.Chr_${i}.msmc2.sorted.vcf.gz
bcftools index ${ind}_NU_Snps.Chr_${i}.msmc2.sorted.vcf.gz
done ;
rm SPI_BJO_ROS_NU_Snps.Chr_${i}.msmc2.sorted.vcf.gz
rm SPI_BJO_ROS_NU_Snps.Chr_${i}.msmc2.sorted.vcf.gz.csi
rm SPI_BJO_ROS_NU_Snps.Chr_${i}.msmc2.vcf.gz
rm SPI_BJO_ROS_NU_Snps.Chr_${i}.nonvariant.vcf.gz
rm SPI_BJO_ROS_NU_Snps.Chr_${i}.nonvariant.vcf.gz.csi
done ;

 
```

```bash
#mappability files according to https://www.ncbi.nlm.nih.gov/pmc/articles/PMC9157554/
module purge
module load seqtk/1.3-GCC-10.3.0

for i in {01..24} ;
do echo ${i} ;
echo "Chr_"${i} > label
seqtk subseq Puffin_ReferenceGenome.NU.MT.fasta label > Puffin_ReferenceGenome.Chr_${i}.fasta ;
done

ml purge
ml Miniconda3/4.9.2
conda activate miniconda3_envs/msmc2

for i in {01..24} ;
do echo ${i} ;
genmap index -F Puffin_ReferenceGenome.Chr_${i}.fasta -I index.Chr_${i}
genmap map -K 35 -E 1 -I index.Chr_${i}/ -O Chr_${i} -r -t -w -bg
awk -F '\t' '$4==1' Chr_${i}.bedgraph | cut -f1-3 > Chr_${i}.mappability.bed 
done

conda deactivate

#vcf and mappability files
for i in {01..24} ;
do echo ${i} ;
cat samples | while read ind ;
do echo $ind ;
zcat ${ind}_NU_Snps.Chr_${i}.msmc2.sorted.vcf.gz | msmc-tools/vcfAllSiteParser.py Chr_${i} ${ind}.Chr_${i}.mask.bed.gz > ${ind}.Chr_${i}.vcf
gzip ${ind}.Chr_${i}.vcf
done 
done

mkdir Results

ml purge
ml Miniconda3/4.9.2
conda activate miniconda3_envs/msmc2

for i in {01..24} ;
do echo ${i} ;
msmc-tools/generate_multihetsep.py --chr Chr_${i} --mask=Chr_${i}.mappability.bed \
--mask=ROS001.Chr_${i}.mask.bed.gz --mask=ROS002.Chr_${i}.mask.bed.gz --mask=ROS003.Chr_${i}.mask.bed.gz --mask=ROS006.Chr_${i}.mask.bed.gz --mask=ROS007.Chr_${i}.mask.bed.gz --mask=ROS008.Chr_${i}.mask.bed.gz \
--mask=SPI001.Chr_${i}.mask.bed.gz --mask=SPI003.Chr_${i}.mask.bed.gz --mask=SPI004.Chr_${i}.mask.bed.gz --mask=SPI011.Chr_${i}.mask.bed.gz --mask=SPI015.Chr_${i}.mask.bed.gz \
ROS001.Chr_${i}.vcf.gz ROS002.Chr_${i}.vcf.gz ROS003.Chr_${i}.vcf.gz ROS006.Chr_${i}.vcf.gz ROS007.Chr_${i}.vcf.gz ROS008.Chr_${i}.vcf.gz \
SPI001.Chr_${i}.vcf.gz SPI003.Chr_${i}.vcf.gz SPI004.Chr_${i}.vcf.gz SPI011.Chr_${i}.vcf.gz SPI015.Chr_${i}.vcf.gz > Results/SPI_ROS.Chr_${i}.multihetsep.msmc-input.txt
done

conda deactivate

```

MSMC2

```bash

cd Results

inds="1 2 3 4 5 6"
dequeue_from_list1() {
  shift;
  list1=$@
}
list1=$inds
for ind1 in $inds ; 
do dequeue_from_list1 $list1 ;
list2=$list1
for ind2 in $list1 ; 
do sbatch MSMC2_1.sh ${ind1} ${ind2} ROS ;
done
done

inds="7 8 9 10 11"
dequeue_from_list1() {
  shift;
  list1=$@
}
list1=$inds
for ind1 in $inds ; 
do dequeue_from_list1 $list1 ;
list2=$list1
for ind2 in $list1 ; 
do sbatch MSMC2_1.sh ${ind1} ${ind2} SPI ;
done
done

### CROSS

inds="1 2 3 4 5 6"
dequeue_from_list1() {
  shift;
  list1=$@
}
dequeue_from_list2() {
  shift;
  list2=$@
}

list1=$inds
for ind1 in $inds ; 
do dequeue_from_list1 $list1 ;
for ind2 in $list1 ; 
do inds2="7 8 9 10 11"
list2=$inds2
for ind3 in $inds2 ; 
do dequeue_from_list2 $list2 ;
for ind4 in $list2 ; 
do sbatch MSMC2_2.sh ${ind1} ${ind2} ${ind3} ${ind4} ;
done
done
done
done

```

MSMC2_1.sh

```bash

module purge
module load MSMC2/2.1.2-linux64

mkdir ${3}.${1}_${2}
cd ${3}.${1}_${2}
cp ../SPI_ROS.Chr_*.multihetsep.msmc-input.txt .

haplo1=$(bc <<<"scale=0; ${1} * 2 - 2 ")
haplo2=$(bc <<<"scale=0; ${1} * 2 - 1 ")
haplo3=$(bc <<<"scale=0; ${2} * 2 - 2 ")
haplo4=$(bc <<<"scale=0; ${2} * 2 - 1 ")

msmc2 -t 24 -s -p 1*2+22*1+1*2+1*6 \
-I ${haplo1},${haplo2},${haplo3},${haplo4} \
-o ${3}.${1}_${2}.msmc2 SPI_ROS.Chr_*.multihetsep.msmc-input.txt

mv ${3}.${1}_${2}.msmc2* ../

cd ../

rm -r ${3}.${1}_${2}

```

MSMC2_2.sh

```bash

module purge
module load MSMC2/2.1.2-linux64

mkdir ${1}_${2}_${3}_${4}
cd ${1}_${2}_${3}_${4}
cp ../SPI_ROS.Chr_*.multihetsep.msmc-input.txt .

haplo1=$(bc <<<"scale=0; ${1} * 2 - 2 ")
haplo2=$(bc <<<"scale=0; ${1} * 2 - 1 ")
haplo3=$(bc <<<"scale=0; ${2} * 2 - 2 ")
haplo4=$(bc <<<"scale=0; ${2} * 2 - 1 ")

haplo5=$(bc <<<"scale=0; ${3} * 2 - 2 ")
haplo6=$(bc <<<"scale=0; ${3} * 2 - 1 ")
haplo7=$(bc <<<"scale=0; ${4} * 2 - 2 ")
haplo8=$(bc <<<"scale=0; ${4} * 2 - 1 ")

msmc2 -t 24 -s -p 1*2+22*1+1*2+1*6 -I \
${haplo1}-${haplo5},${haplo1}-${haplo6},${haplo1}-${haplo7},${haplo1}-${haplo8},\
${haplo2}-${haplo5},${haplo2}-${haplo6},${haplo2}-${haplo7},${haplo2}-${haplo8},\
${haplo3}-${haplo5},${haplo3}-${haplo6},${haplo3}-${haplo7},${haplo3}-${haplo8},\
${haplo4}-${haplo5},${haplo4}-${haplo6},${haplo4}-${haplo7},${haplo4}-${haplo8} \
-o ROS_SPI.${1}_${2}_${3}_${4}.msmc2 SPI_ROS.Chr_*.multihetsep.msmc-input.txt

mv ROS_SPI.${1}_${2}_${3}_${4}.msmc* ../

cd ../

rm -r ${1}_${2}_${3}_${4}

```

Transfer and Quick plottin in R

```bash

#COMBINED

ml purge
ml Miniconda3/4.9.2
conda activate msmc2

ls | grep "ROS_SPI" | grep "final" | while read line ;
do echo ${line} ;
ind1=$(echo ${line} | awk -F '_' '{print $2}' | awk -F '.' '{print $2}')
ind2=$(echo ${line} | awk -F '_' '{print $3}')
ind3=$(echo ${line} | awk -F '_' '{print $4}')
ind4=$(echo ${line} | awk -F '_' '{print $5}' | awk -F '.' '{print $1}')
../msmc-tools/combineCrossCoal.py ./ROS_SPI.${ind1}_${ind2}_${ind3}_${ind4}.msmc2.final.txt ./ROS.${ind1}_${ind2}.msmc2.final.txt \
./SPI.${ind3}_${ind4}.msmc2.final.txt > ./ROS_SPI.${ind1}_${ind2}_${ind3}_${ind4}.combined.msmc2.final.txt
done


head -1 SPI.10_11.msmc2.final.txt > header1
echo "Sample" > header2
paste header1 header2 > header

touch SPI.ALL.msmc2.final.txt
cat header >> SPI.ALL.msmc2.final.txt
ls * | grep "SPI." | grep -v "ROS" | grep -v "ALL" | while read line ;
do echo ${line} ;
sample=$(echo ${line} | awk -F '.' '{print $2}')
cat ${line} | tail -n+2 | awk -v sam=${sample} '{print $1"\t"$2"\t"$3"\t"$4"\t"sam}' >> SPI.ALL.msmc2.final.txt ;
done

touch ROS.ALL.msmc2.final.txt
cat header >> ROS.ALL.msmc2.final.txt
ls * | grep "ROS." | grep -v "SPI" | grep -v "ALL" | while read line ;
do echo ${line} ;
sample=$(echo ${line} | awk -F '.' '{print $2}')
cat ${line} | tail -n+2 | awk -v sam=${sample} '{print $1"\t"$2"\t"$3"\t"$4"\t"sam}' >> ROS.ALL.msmc2.final.txt ;
done

head -1 ROS_SPI.2_4_8_11.combined.msmc2.final.txt > header3
echo "Sample" > header4
paste header3 header4 > header5
touch ROS_SPI.ALL.combined.msmc2.final.txt
cat header5 >> ROS_SPI.ALL.combined.msmc2.final.txt
ls * | grep "ROS_SPI" | grep "combined" | grep -v "ALL" | while read line ;
do echo ${line} ;
sample=$(echo ${line} | awk -F '.' '{print $2}')
cat ${line} | tail -n+2 | awk -v sam=${sample} '{print $1"\t"$2"\t"$3"\t"$4"\t"$5"\t"$6"\t"sam}' >> ROS_SPI.ALL.combined.msmc2.final.txt ;
done

#Rscript(s)

```

need to do Bjørnøya as well

```bash

ml purge
ml Miniconda3/4.9.2
conda activate miniconda3_envs/msmc2

for i in {01..24} ;
do echo ${i} ;
../msmc-tools/generate_multihetsep.py --chr Chr_${i} --mask=../Chr_${i}.mappability.bed \
--mask=../BJO001.Chr_${i}.mask.bed.gz --mask=../BJO003.Chr_${i}.mask.bed.gz --mask=../BJO005.Chr_${i}.mask.bed.gz --mask=../BJO006.Chr_${i}.mask.bed.gz --mask=../BJO007.Chr_${i}.mask.bed.gz --mask=../BJO008.Chr_${i}.mask.bed.gz \
../BJO001.Chr_${i}.vcf.gz ../BJO003.Chr_${i}.vcf.gz ../BJO005.Chr_${i}.vcf.gz ../BJO006.Chr_${i}.vcf.gz ../BJO007.Chr_${i}.vcf.gz ../BJO008.Chr_${i}.vcf.gz > Results/BJO.Chr_${i}.multihetsep.msmc-input.txt
done

cd Results
inds="1 2 3 4 5 6"
dequeue_from_list1() {
  shift;
  list1=$@
}
list1=$inds
for ind1 in $inds ; 
do dequeue_from_list1 $list1 ;
list2=$list1
for ind2 in $list1 ; 
do sbatch MSMC2_1.sh ${ind1} ${ind2} BJO ;
done
done

```

MSMC2_1.sh

```bash

module purge
module load MSMC2/2.1.2-linux64

mkdir ${3}.${1}_${2}
cd ${3}.${1}_${2}
cp ../BJO.Chr_*.multihetsep.msmc-input.txt .

haplo1=$(bc <<<"scale=0; ${1} * 2 - 2 ")
haplo2=$(bc <<<"scale=0; ${1} * 2 - 1 ")
haplo3=$(bc <<<"scale=0; ${2} * 2 - 2 ")
haplo4=$(bc <<<"scale=0; ${2} * 2 - 1 ")

msmc2 -t 24 -s -p 1*2+22*1+1*2+1*6 \
-I ${haplo1},${haplo2},${haplo3},${haplo4} \
-o ${3}.${1}_${2}.msmc2 BJO.Chr_*.multihetsep.msmc-input.txt

mv ${3}.${1}_${2}.msmc2* ../

cd ../

rm -r ${3}.${1}_${2}

```

```bash

touch BJO.ALL.msmc2.final.txt
cat header >> BJO.ALL.msmc2.final.txt
ls * | grep "BJO." | grep -v "SPI" | grep -v "ROS" | grep -v "ALL" | while read line ;
do echo ${line} ;
sample=$(echo ${line} | awk -F '.' '{print $2}')
cat ${line} | tail -n+2 | awk -v sam=${sample} '{print $1"\t"$2"\t"$3"\t"$4"\t"sam}' >> BJO.ALL.msmc2.final.txt ;
done

```