# RFmix 


RFmix

```bash
mkdir RFmix
cd RFmix

cp SPI_BJO_ROS.NoRelate.Chr_*.phased.vcf.gz* .
cp Globetrotter/IndividualsOverview.txt .
cp Globetrotter/Chr_*.chromo.edit.rec .

#Genetic map
for i in {01..24} ;
do echo ${i} ;
echo -e "#chromosome\tposition\tGenetic_Map(cM)" > Gmap.Chr_${i}
cat Chr_${i}.chromo.edit.rec | tail -n+2 | awk -v chr=${i} '($3=100*$1*$2) {print "Chr_"chr"\t"$1"\t"$3}' >> Gmap.Chr_${i}
done

echo -e "#chromosome\tposition\tGenetic_Map(cM)" > GeneticMap.All
for i in {01..24} ;
do echo ${i} ;
cat Gmap.Chr_${i} | tail -n+2 >> GeneticMap.All
done

cat IndividualsOverview.txt | grep -v "BJO" | cut -f1 >> Reference.samples
cat IndividualsOverview.txt | grep "BJO" | cut -f1 >> Query.samples
cat IndividualsOverview.txt | cut -f1-2 | grep -v "BJO" > Reference_samples.map

module purge
module load BCFtools/1.9-intel-2018b

for i in {01..24} ;
do echo ${i} ;
bcftools view -S Reference.samples -o SPI_ROS.Chr_${i}.phased.vcf SPI_BJO_ROS.NoRelate.Chr_${i}.phased.vcf.gz
done
bcftools concat -O b -o SPI_ROS.All.phased.bcf.gz SPI_ROS.Chr_*.phased.vcf
bcftools sort -O b -o SPI_ROS.All.phased.sorted.bcf.gz SPI_ROS.All.phased.bcf.gz
bcftools index SPI_ROS.All.phased.sorted.bcf.gz
rm SPI_ROS.Chr_*.phased.vcf
rm SPI_ROS.All.phased.bcf.gz

for i in {01..24} ;
do echo ${i} ;
bcftools view -S Query.samples -o BJO.Chr_${i}.phased.vcf SPI_BJO_ROS.NoRelate.Chr_${i}.phased.vcf.gz
done
bcftools concat -O b -o BJO.All.phased.bcf.gz BJO.Chr_*.phased.vcf
bcftools sort -O b -o BJO.All.phased.sorted.bcf.gz BJO.All.phased.bcf.gz
bcftools index BJO.All.phased.sorted.bcf.gz
rm BJO.Chr_*.phased.vcf
rm BJO.All.phased.bcf.gz

mkdir Results/

for i in {01..24} ;
do echo ${i} ;
sbatch RFmix.sh ${i}
done

```

RFmix.sh

```bash

module purge
module load BCFtools/1.9-intel-2018b

rfmix -f BJO.All.phased.sorted.bcf.gz \
-r SPI_ROS.All.phased.sorted.bcf.gz \
-m Reference_samples.map \
-g GeneticMap.All \
-o BJO_RFmix_Chr_${1} \
-e 50 \
--reanalyze-reference \
--chromosome=Chr_${1}

mv BJO_RFmix_Chr_${1}* Results/

```

```bash

touch BJO_RFmix_All.rfmix.Q
for i in {01..24} ;
do echo ${i} ;
grep -v "#" BJO_RFmix_Chr_${i}.rfmix.Q | head -6 | awk -v chr=${i} '{print $1"\t"$2"\t"$3"\t""chr"chr}' >> BJO_RFmix_All.rfmix.Q
done 

head -2 BJO_RFmix_Chr_01.msp.tsv | tail -1 > BJO_RFmix_All.msp.tsv
for i in {01..24} ;
do echo ${i} ;
tail -n+3 BJO_RFmix_Chr_${i}.msp.tsv >> BJO_RFmix_All.msp.tsv
done 

nano BJO_RFmix_All.msp.tsv #remove hash

#Rscript for plots
cat samples | awk -F '_' '{print $1}' | sort | uniq | while read sample ;
do echo $sample  ;
Rscript Ideogram.R ${sample} ;
done

#(Window) Length
#make bed files in R
cat samples | awk -F '_' '{print $1}' | sort | uniq | while read sample ;
do echo $sample  ;
Rscript Bedfiles.R ${sample} ;
done

#non window based
touch All.combined.bed
cat samples | awk -F '_' '{print $1}' | sort | uniq | while read sample ;
do echo $sample  ;
cat ${sample}_1.bed | awk -v var=${sample} '{print $1"\t"$2"\t"$3"\t"$4"\t"$5"\t"var}' >> All.combined.bed ;
cat ${sample}_2.bed | awk -v var=${sample} '{print $1"\t"$2"\t"$3"\t"$4"\t"$5"\t"var}' >> All.combined.bed ;
done

#make windows of 100kb
awk -v OFS='\t' {'print $1,$2'} Puffin_ReferenceGenome.NU.FINAL.fasta.fai > Puffin_ReferenceGenome.NU.FINAL.txt
bedtools makewindows -g Puffin_ReferenceGenome.NU.FINAL.txt -w 100000 > Puffin_ReferenceGenome.NU.FINAL.windows

cat Puffin_ReferenceGenome.NU.FINAL.windows | sed 's/Chr_/chr/g' > Chr_All.windows

cat samples | while read sample ;
do echo $sample ;
Line=$(grep -n -w ${sample} samples | awk -F ':' '{print $1}')
col1=$(bc <<<"scale=0; ${Line} * 3 - 2 + 1 ")
col2=$(bc <<<"scale=0; ${Line} * 3 - 1 + 1 ")
col3=$(bc <<<"scale=0; ${Line} * 3 + 1 ")
cut -d ' ' -f1,${col1},${col2},${col3} Positions_haplo_intro_Chr_${i} | grep -v -w "NA" | tail -n+2 | awk '{print $1"\t"$3"\t"$2"\t""+""\t"$4}' | sort -n -k2,2 | sed -r "s/^${i}/chr${i}/g" > Chr_${i}_${sample}.bed ;
done

bedtools intersect -a Chr_All.windows \
-b BJO001_1.bed BJO001_2.bed BJO003_1.bed BJO003_2.bed \
BJO005_1.bed BJO005_2.bed BJO006_1.bed BJO006_2.bed \
BJO007_1.bed BJO007_2.bed BJO008_1.bed BJO008_2.bed \
-wa -wb -filenames > Allsamples_intersect.result

cut -f5-9 Allsamples_intersect.result | bedtools sort -i - > Allsamples_intersect.results.bed
bedtools map -a Chr_All.windows -b Allsamples_intersect.results.bed -c 5 -o mean > Allsamples_meanLength.bed

for i in {01..24} ;
do echo ${i} ;
grep -w "chr"${i} Chr_All.windows > Chr_${i}.windows ;
cat samples | awk -F '_' '{print $1}' | sort | uniq | while read sample ;
do echo $sample  ;
grep -w "chr"${i} ${sample}_1.bed > Chr_${i}_${sample}_1.bed
grep -w "chr"${i} ${sample}_2.bed > Chr_${i}_${sample}_2.bed
done
done

for i in {01..24} ;
do echo ${i} ;
bedtools coverage -a Chr_${i}.windows \
-b Chr_${i}_BJO001_1.bed Chr_${i}_BJO001_2.bed Chr_${i}_BJO003_1.bed Chr_${i}_BJO003_2.bed \
Chr_${i}_BJO005_1.bed Chr_${i}_BJO005_2.bed Chr_${i}_BJO006_1.bed Chr_${i}_BJO006_2.bed \
Chr_${i}_BJO007_1.bed Chr_${i}_BJO007_2.bed Chr_${i}_BJO008_1.bed Chr_${i}_BJO008_2.bed \
-d > Chr_${i}.allsamples_coverage.result

bedtools map -a Chr_${i}.windows -b Chr_${i}.allsamples_coverage.result -c 5 -o mean > Chr_${i}.allsamples_meanCount.bed
rm Chr_${i}.allsamples_coverage.result
done

touch Allsamples_meanCount.bed
for i in {01..24} ;
do echo ${i} ;
cat Chr_${i}.allsamples_meanCount.bed >> Allsamples_meanCount.bed
done

grep -v "chrUnplaced" Allsamples_meanLength.bed | grep -v "chrZ" | grep -v "chrW" > Allsamples_meanLength.edit.bed
cut -f4 Allsamples_meanCount.bed | paste Allsamples_meanLength.edit.bed - > Allsamples_meanLength_Count.bed

```