# Quality Check


ANGSD - QualityCheck

```bash
rm slurm*

# listing files for ANGSD to work on:
ls *.bam | grep -v "IOM001" | grep -v "RAZ_Ancestral" > bams
nano bams #remove the blanks + SPI related ind. (SPI002)

# get reference for ANGSD commands
cp ../Paleomix/Reference/Puffin_ReferenceGenome.NU.MT.FINAL.fasta .

#Check these parameters, or run without it
#-C 50
#-baq 1/2
# BAQ 2 (or "extended BAQ") as far as I understand should be more sensitive at the cost of accuracy. So more detected SNP's but
# also more false positive SNP's as compared to BAQ 1. I currently use BAQ 2, because my impression has been that BAQ 1 could
# lead to too many false negatives. I also get a much better looking site frequency spectrum with BAQ 2 as compared to BAQ 1.
# The default BAQ is oversensitive to clustered SNPs. The extended BAQ fixes this problem. In the samtools repository, -E
# is the default behavior (which means future versions of samtools will also use -E by default). That said, for human, the
# difference between -B and -E is mostly a couple of percent. I guess the SNP density is pretty high in your sample.
# The huge difference between -B and -E implies the indel density is pretty high, further implying the SNP density is also high.

sbatch Puffin_ANGSD_QC.sh

```

nano Puffin_ANGSD_QC.sh

```bash


# angsd settings:
# -minMapQ 25 : only highly unique mappings (prob of erroneous mapping = 0.5%)
# -maxDepth : highest total depth (sum over all samples) to assess; set to ca . 2x Ave.Cov. x number of samples
FILTERS="-uniqueOnly 1 -remove_bads 1 -minMapQ 25 -maxDepth 1100 -checkBamHeaders 1 -C 50 -baq 2 -rmTrans 1 "

# T O   D O :
TODO="-doQsDist 1 -doDepth 1 -doCounts 1 -dumpCounts 2 -doMajorMinor 1"

# Run command:
module purge 
module load SAMtools/1.9-GCC-8.2.0-2.31.1
samtools faidx Puffin_ReferenceGenome.NU.MT.FINAL.fasta

module load angsd/0.931-GCC-8.2.0-2.31.1

Threads="12"

angsd -b bams -GL 1 $FILTERS $TODO -P ${Threads} -ref Puffin_ReferenceGenome.NU.MT.FINAL.fasta -r Chr_01: -out PuffQC

```

```bash
cd AncientPuffins/ANGSD

gunzip PuffQC.counts.gz
#subsample counts for every 10th SNP starting after header (reducing file size so R can handle - should be about 20M SNPs)
head -1 PuffQC.counts > PuffQC.header
awk 'NR % 10 == 2' PuffQC.counts > PuffQC.counts2
cat PuffQC.header PuffQC.counts2 > PuffQC.counts3
wc -l PuffQC.counts3

rm PuffQC.counts2
rm PuffQC.header
gzip PuffQC.counts3
gzip PuffQC.counts

gunzip PuffQC.counts3.gz
Rscript ./plotQC.R PuffQC > qranks #warning messages are fine #need to adjust the number of individuals/columns and n's for last graph
gzip PuffQC.counts3

# percentages of sites with coverage >5x in each sample, from worst to best:
cat qranks
# look at PuffQC.pdf for more details

```