# Main ANGSD


```bash

#make result directory
mkdir GenoLikeResult

echo -e "Chr_01:\nChr_02:\nChr_03:\nChr_04:\nChr_05:\nChr_06:\nChr_07:\nChr_08:\nChr_09:\nChr_10:\nChr_11:\nChr_12:\nChr_13:\nChr_14:\nChr_15:\nChr_16:\nChr_17:\nChr_18:\nChr_19:\nChr_20:\nChr_21:\nChr_22:\nChr_23:\nChr_24:\nChr_W:\nChr_Z:\nChr_Unplaced:" > ChromosomeList

#run Puffin_ANGSD_main.sh
cat ChromosomeList | grep -v "Chr_Z" | grep -v "Chr_W" | grep -v "Chr_Unplaced" | while read i ;
do echo ${i} ;
sbatch Puffin_ANGSD_main.sh ${i} ;
done

```

nano Puffin_ANGSD_main.sh

```bash


mkdir temp.${1}
cp -r Puffin_ReferenceGenome.NU.MT.FINAL.fasta temp.${1}/ #reference genome
cp bams_good temp.${1}/ #list of bamfiles
cp -r *.Puffin_ReferenceGenome.NU.sorted.reheadered.bam temp.${1}/
cp -r *.Puffin_ReferenceGenome.NU.sorted.reheadered.bam.bai temp.${1}/

cd temp.${1}

Chromosome=$(echo ${1} | awk -F ':' '{print $1}')
#script is called per chromosome and file names will be named with this EnvVariable

# F I L T E R S :
FILTERS="-uniqueOnly 1 -remove_bads 1 -minMapQ 30 -minQ 30 -dosnpstat 1 -C 50 -baq 2 -checkBamHeaders 1 -doHWE 1 -HWE_pval 1e-2 -sb_pval 1e-5 -hetbias_pval 1e-5 -skipTriallelic 1 -minInd 95 -snp_pval 1e-6 -minMaf 0.05 -setMaxDepth 875 -setMinDepth 379 -rmTrans 1"

# T O   D O :
TODO="-doMajorMinor 1 -doMaf 1 -doCounts 1 -doGeno 8 -doPost 1 -doGlf 2"

# Run command:
module purge 
module load SAMtools/1.9-GCC-8.2.0-2.31.1
samtools faidx Puffin_ReferenceGenome.NU.MT.FINAL.fasta

module purge 
module load angsd/0.935-GCC-10.2.0

Threads="16"

angsd -b bams_good -GL 1 $FILTERS $TODO -P ${Threads} -ref Puffin_ReferenceGenome.NU.MT.FINAL.fasta -r ${1} -out PuffinAngsd_${Chromosome}

cp -r PuffinAngsd_${Chromosome}* ../GenoLikeResult/

cd ../

rm -r temp.${1}

```

SUMMARY

```bash
cd /cluster/work/users/oliverke/AncientPuffins/ANGSD/GenoLikeResult
zcat PuffinAngsd_Chr_01.mafs.gz > PuffinAngsd_Combined.mafs

for i in 02 03 04 05 06 07 08 09 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 ;
do zcat PuffinAngsd_Chr_${i}.mafs.gz | tail -n +2 >> PuffinAngsd_Combined.mafs ;
done ;
gzip PuffinAngsd_Combined.mafs
zcat PuffinAngsd_Combined.mafs.gz | tail -n +2 | wc -l


```