# LD pruning

ngsLD

```bash
mkdir LD
cd LD

cp ../ChromosomeList .
sed -i 's/://g' ChromosomeList

cp ChromosomeList ChromosomeList_new
nano ChromosomeList_new

cat ChromosomeList_new | while read i ;
do echo ${i} ;
sbatch Puffin_ngsLD_decay.sh ${i} ;
done


```

nanoÂ Puffin_ngsLD_decay.sh

```bash

mkdir temp${1}
cd temp${1}

cp -r GenoLikeResult/PuffinAngsd_${1}.mafs.gz .
cp -r GenoLikeResult/PuffinAngsd_${1}.beagle.gz .

zcat PuffinAngsd_${1}.mafs.gz | cut -f 1,2 | tail -n +2 > PuffinAngsd_${1}.pos
NS=$(cat PuffinAngsd_${1}.pos | wc -l)
echo "Number of sites = "$NS

#ngsLD
module load ngsLD/191108-GCC-8.2.0-2.31.1

ngsLD --n_threads 16 --geno PuffinAngsd_${1}.beagle.gz --pos PuffinAngsd_${1}.pos --probs --n_ind 98 --n_sites $NS --max_kb_dist 50 | gzip --best > PuffinAngsd_${1}.ld.gz

#DecayPlot
#subsample output (5% of lines) for Decay Plot
zcat PuffinAngsd_${1}.ld.gz | awk 'rand()<0.05' | gzip --best > PuffinAngsd_${1}.ld_sampled.gz

cp -r PuffinAngsd_${1}.pos ../
cp -r PuffinAngsd_${1}.ld.gz ../
cp -r PuffinAngsd_${1}.ld_sampled.gz ../

cd ../

rm -r temp${1}/

```

Pruning

```bash
cat ChromosomeList_new | while read i ;
do echo ${i} ;
sbatch Puffin_LDpruning.sh ${i} ;
done


```

nano Puffin_LDpruning.sh

```bash

mkdir temp${1}
cd temp${1}

cp -r GenoLikeResult/PuffinAngsd_${1}.beagle.gz .
cp -r LD/PuffinAngsd_${1}.ld.gz .

module load OrthoMCL/2.0.9-intel-2018b-Perl-5.28.0

gunzip PuffinAngsd_${1}.ld.gz
gunzip PuffinAngsd_${1}.beagle.gz
tail -n +2 PuffinAngsd_${1}.beagle > PuffinAngsd_${1}.beagle.edit

mcl <(cut -f1,2,7 PuffinAngsd_${1}.ld | awk '$3>0.2') --abc -I 2.0 -o PuffinAngsd_${1}.ld.mcl
# All SNPs with a LD > 0.2 are considered here. You might want to change it. The file DOM2.$CHR.LD.mcl will contain groups of
# linked SNPs. I just took the most central one, as representative of the haplotype block.

#non-cluster snps
awk '{print $1"\t"$2"\t"$7}' PuffinAngsd_${1}.ld | awk '$3>0.2' | awk '{print $1}' > tmp1
awk '{print $1"\t"$2"\t"$7}' PuffinAngsd_${1}.ld | awk '$3>0.2' | awk '{print $2}' > tmp2
cat tmp1 tmp2 | sort | uniq > cluster.snps
sed -i 's/:/_/g' cluster.snps
rm tmp1
rm tmp2
LANG=C fgrep -v -w -f cluster.snps PuffinAngsd_${1}.beagle.edit > tmp1

#rep snps of cluster
touch Rep.snps
TotalLines=$(cat PuffinAngsd_${1}.ld.mcl | wc -l)
echo "Number of total lines = "$TotalLines
for line in $(seq 1 1 $TotalLines) ;
do ColumnNumbers=$(head -${line} PuffinAngsd_${1}.ld.mcl | tail -1 | awk '{print NF}' | sort -nu | tail -n 1) ;
AlmostMiddleColumn=$(bc <<<"scale=0; $ColumnNumbers / 2" ) ;
let "MiddleColumn=$AlmostMiddleColumn+1" ;
Scaffold=$(head -${line} PuffinAngsd_${1}.ld.mcl | tail -1 | awk -v VAR=$MiddleColumn '{print $VAR}') ;
echo $Scaffold >> Rep.snps ;
done ;

cat Rep.snps | sort | uniq > Rep.snps.uniq
sed -i 's/:/_/g' Rep.snps.uniq

LANG=C fgrep -w -f Rep.snps.uniq PuffinAngsd_${1}.beagle.edit > tmp2

#merging
cat tmp1 tmp2 | sort -t '_' -n -k4,4 > PuffinAngsd_${1}.beagle.tmp
head -1 PuffinAngsd_${1}.beagle | cat - PuffinAngsd_${1}.beagle.tmp > PuffinAngsd_${1}.beagle.ldpruned
gzip PuffinAngsd_${1}.beagle.ldpruned

cp -r PuffinAngsd_${1}.beagle.ldpruned.gz ../
cd ../

rm -r temp${1}/

```

Combine chromosomes

```bash

zcat PuffinAngsd_Chr_01.beagle.ldpruned.gz > PuffinAngsd_Chr_Autosomes.beagle.ldpruned

for i in 02 03 04 05 06 07 08 09 10 11 12 13 14 15 16 17 18 19 20 21 22 23 24 ;
do zcat PuffinAngsd_Chr_${i}.beagle.ldpruned.gz | tail -n +2 >> PuffinAngsd_Chr_Autosomes.beagle.ldpruned ;
done ;
gzip PuffinAngsd_Chr_Autosomes.beagle.ldpruned
zcat PuffinAngsd_Chr_Autosomes.beagle.ldpruned.gz | tail -n +2 | wc -l


```