# SFS Analyses



```bash
mkdir SFS
cd SFS

cp ../Puffin_ReferenceGenome.NU.MT.FINAL.fasta .
cp ../ChromosomeList .

grep -v "Z" ChromosomeList | grep -v "Unplaced" | grep -v "25" | grep -v "W" > ChromosomeList.edit

# make neccessary pop files, remove related ind.
ls ../*.bam | grep -v "OK" |Â awk -F '/' '{print $2}' > bams_good

#running a global ANGSD run for all sites to get sites info
# make result directory
mkdir AngsdResults
#run Puffin_ANGSD1_SFS.sh
cat ChromosomeList.edit | while read i ;
do echo ${i} ;
sbatch Puffin_ANGSD_SFS1.sh ${i} ;
done

```

nano Puffin_ANGSD_SFS1.sh

```bash


mkdir temp.${1}
cp Puffin_ReferenceGenome.NU.MT.FINAL.fasta temp.${1}/
cp bams_good temp.${1}/
cp -r ../*.Puffin_ReferenceGenome.NU.sorted.reheadered.bam temp.${1}/
cp -r ../*.Puffin_ReferenceGenome.NU.sorted.reheadered.bam.bai temp.${1}/

cd temp.${1}/

module purge
module load SAMtools/1.9-GCC-8.2.0-2.31.1
samtools faidx Puffin_ReferenceGenome.NU.MT.FINAL.fasta

module purge 
module load angsd/0.935-GCC-10.2.0

Chromosome=$(echo ${1} | awk -F ':' '{print $1}')

# F I L T E R S :
# Filters are different because SFS needs rare alleles. Hence no MAF or snp pvalue. Also no LD pruning
FILTERS="-uniqueOnly 1 -remove_bads 1 -minMapQ 30 -minQ 30 -dosnpstat 1 -C 50 -baq 2 -checkBamHeaders 1 -doHWE 1 -sb_pval 1e-5 -hetbias_pval 1e-5 -skipTriallelic 1 -minInd 95 -setMaxDepth 875 -setMinDepth 379 -rmTrans 1"

# T O   D O :
TODO="-doMajorMinor 1 -doMaf 1 -doCounts 1 -doGeno 3 -doPost 2"

# Run command:

Threads="16"

angsd -b bams_good -GL 1 $FILTERS $TODO -P ${Threads} -ref Puffin_ReferenceGenome.NU.MT.FINAL.fasta -r ${1} -out PuffinAngsd_SFS_${Chromosome}

mv PuffinAngsd_SFS_${Chromosome}.snpStat.gz ../AngsdResults/

cd ../
rm -r temp.${1}/

```

```bash

rm slurm*
cd AngsdResults
# merging all sites
zcat PuffinAngsd_SFS_Chr_01.snpStat.gz > sfsSites.snpStat


sbatch Concat.sh 
```

nano Concat.sh

```bash


for i in {02..24} ;
do zcat PuffinAngsd_SFS_Chr_${i}.snpStat.gz | tail -n+2 >> sfsSites.snpStat ;
done ;

```

```bash

rm slurm*
# how many sites # took 5 min
tail -n+2 sfsSites.snpStat | wc -l 
#zipping it
gzip sfsSites.snpStat

# extracting, sorting, and indexing list of sites to make SFS from 
# filtering out sites where heterozygote counts comprise more than 50% of all counts (likely lumped paralogs)
zcat sfsSites.snpStat.gz | awk '($3+$4+$5+$6)>0' | awk '($12+$13+$14+$15)/($3+$4+$5+$6)<0.5' | cut -f 1,2 > sites2do

wc -l sites2do


cp sites2do ../

cd ..

# make neccessary pop files

#Thule only 3 large ind. for "large" pop
#SPH only the ones from 1900 + other ones are only 3 from 1860s
#Plus a SPH combined
#not SPH004 + 008 because dwarfs

grep "GUL" bams_good >> GUL.samples
grep "GAN" bams_good >> GAN.samples
cat GUL.samples GAN.samples >> CAN.samples
grep "IOM" bams_good >> IOM.samples
grep "ROS" bams_good >> ROS.samples
grep "WES" bams_good >> WES.samples
grep "BRE" bams_good >> BRE.samples
grep "GRI" bams_good >> GRI.samples
grep "PAP" bams_good >> PAP.samples
grep "FAR" bams_good >> FAR.samples
grep "HOR" bams_good >> HOR.samples
cat WES.samples BRE.samples GRI.samples PAP.samples FAR.samples HOR.samples ROS.samples >> MAI.samples
cat WES.samples BRE.samples GRI.samples PAP.samples >> ICE.samples
grep "SPI" bams_good >> SPI.samples #remove SPI002
grep "ROH" bams_good >> ROH.samples
grep "BJO" bams_good >> BJO.samples
grep "BJH" bams_good >> BJH.samples
grep "SPH005" bams_good >> SPH_Young.samples
grep "SPH006" bams_good >> SPH_Young.samples
grep "SPH007" bams_good >> SPH_Young.samples
grep "SPH009" bams_good >> SPH_Young.samples
grep "SPH010" bams_good >> SPH_Young.samples
grep "SPH001" bams_good >> SPH_Old.samples
grep "SPH002" bams_good >> SPH_Old.samples
grep "SPH003" bams_good >> SPH_Old.samples
cat SPH_Young.samples SPH_Old.samples >> SPH.samples
grep "THU008" bams_good >> THU_Large.samples
grep "THU010" bams_good >> THU_Large.samples
grep "THU016" bams_good >> THU_Large.samples

mkdir WinSFS_Results

# Estimate SFS from sites2do (pop-wise)
# do the same with alt script - more juice for chr 1 2 3

for pop in CAN IOM MAI SPI BJO THU_Large ROS BJH SPH ROH SPH_Old SPH_Young ;
do echo $pop
sbatch SFS_Angsd.sh ${pop} ;
sbatch SFS_AngsdAlt.sh ${pop} ;
done

```

nano SFS_Angsd.sh ( SFS_AngsdAlt.sh with 20G)

```bash


mkdir temp.${1}_${SLURM_ARRAY_TASK_ID}

cp -r Puffin_ReferenceGenome.NU.MT.FINAL.fasta temp.${1}_${SLURM_ARRAY_TASK_ID}/ #reference genome

cp ChromosomeList.edit temp.${1}_${SLURM_ARRAY_TASK_ID}/ #list of chromosomes

cp ${1}.samples temp.${1}_${SLURM_ARRAY_TASK_ID}/

cd temp.${1}_${SLURM_ARRAY_TASK_ID}/

# Run command:

cat ${1}.samples | while read line ;
do echo "../../"${line} >> ${1}.samples.edit ;
done

module purge 
module load SAMtools/1.9-GCC-8.2.0-2.31.1
samtools faidx Puffin_ReferenceGenome.NU.MT.FINAL.fasta

module purge 
module load angsd/0.935-GCC-10.2.0

Threads="1"
GENOME_ANC=Puffin_ReferenceGenome.NU.MT.FINAL.fasta
GENOME_REF=Puffin_ReferenceGenome.NU.MT.FINAL.fasta
Population=$(echo ${1})

TODO="-doSaf 1"

if [[ ${SLURM_ARRAY_TASK_ID} -gt 9 ]]
then
	ChromoNo=$(echo ${SLURM_ARRAY_TASK_ID} )
else
	ChromoNo=$(echo "0"${SLURM_ARRAY_TASK_ID} )
fi

Chromosome=$(cat ChromosomeList.edit | awk -F ':' '{print $1}' | grep -w Chr_${ChromoNo} )
ChromosomeEdit=$(echo ${Chromosome}":")
grep -w Chr_${ChromoNo} ../sites2do > sites_Chromosome${ChromoNo}_2do

sleep 120

angsd sites index sites_Chromosome${ChromoNo}_2do

sleep 120 

angsd -r ${ChromosomeEdit} -sites sites_Chromosome${ChromoNo}_2do -b ${1}.samples.edit -GL 1 -P ${Threads} -anc ${GENOME_ANC} -ref ${GENOME_REF} $TODO -out ${Population}_Chromosome${ChromoNo}

mv sites_Chromosome${ChromoNo}_2do ${Population}.sites_Chromosome${ChromoNo}_2do
cp -r ${Population}* ../WinSFS_Results/
rm -r ../WinSFS_Results/${Population}.sites_Chromosome${ChromoNo}_2do

cd ..
rm -r temp.${1}_${SLURM_ARRAY_TASK_ID}/

```

WINSFS + realSFS


```bash
cd WinSFS_Results/

for pop in CAN IOM MAI SPI BJO THU_Large ROS BJH SPH ROH SPH_Old SPH_Young ;
do echo $pop
sbatch SFS_Calc1.sh ${pop};
done

```

nano SFS_Calc1.sh

```bash


# Run command:
module load angsd/0.931-GCC-8.2.0-2.31.1
Threads="16"
Population=$(echo ${1})

realSFS cat ${Population}_Chromosome*.saf.idx -outnames ${Population} 
realSFS ${Population}.saf.idx -P ${Threads} -fold 1 > ${Population}.sfs 

rm ${Population}_Chromosome*.saf.gz
rm ${Population}_Chromosome*.saf.idx
rm ${Population}_Chromosome*.saf.pos.gz
rm ${Population}_Chromosome*.arg 

```

winsfs

```bash
for pop in CAN IOM MAI SPI BJO THU_Large ROS BJH SPH ROH SPH_Old SPH_Young ;
do echo $pop
sbatch Winsfs.sh ${pop};
done

```

Winsfs.sh

```bash


# Run command:
Threads="16"
Population=$(echo ${1})

winsfs -v -t ${Threads} ${Population}.saf.idx > ${Population}.winsfs.sfs

head -1 ${Population}.winsfs.sfs > header_${Population}

Nbin=$(tail -1 ${Population}.winsfs.sfs | tr ' ' '\n' | wc -l)
Nbin_half=$(bc <<<"scale=0; ( ${Nbin} + 1 ) / 2")
Nbin_edit=$(echo $Nbin)

while [[ $Nbin_edit -gt $Nbin_half ]] 
do
	Line_one=$(bc <<<"scale=0; ${Nbin} - ${Nbin_edit} + 1")
	One=$(tail -1 ${Population}.winsfs.sfs | tr ' ' '\n' | head -${Line_one} | tail -1)
	Two=$(tail -1 ${Population}.winsfs.sfs | tr ' ' '\n' | head -${Nbin_edit} | tail -1) 
	Number=$(bc <<<"scale=5; ${One} + ${Two}")
	echo $Number >> tmp_${Population}
	Nbin_edit=$(bc <<<"scale=1; ${Nbin_edit} - 1")
done
tail -1 ${Population}.winsfs.sfs | tr ' ' '\n' | head -${Nbin_edit} | tail -1 >> tmp_${Population}

while [[ $Nbin_half -gt 1 ]] 
do 
	echo ${Nbin_half}
	echo "0.000000" >> tmp_${Population}
	Nbin_half=$(bc <<<"scale=1; ${Nbin_half} - 1")
done

cat tmp_${Population} | tr '\n' ' ' > body_${Population}
echo -e "\n" >> body_${Population}

cat header_${Population} body_${Population} > ${Population}.winsfs_folded.sfs

rm tmp_${Population}
rm header_${Population}
rm body_${Population}

```

THEN Theta

```bash
cd /cluster/work/users/oliverke/AncientPuffins/ANGSD/SFS
rm WinSFS_Results/slurm*

mkdir Theta_Results
cd Theta_Results

for pop in CAN IOM MAI SPI BJO THU_Large ROS BJH SPH ROH SPH_Old SPH_Young ;
do echo $pop
sbatch SFS_Theta.sh ${pop} 
done

```

nano SFS_Theta.sh 

```bash

Population=$(echo ${1})

mkdir temp.${Population}

cp ../WinSFS_Results/${Population}.winsfs_folded.sfs temp.${Population}/
cp ../WinSFS_Results/${Population}.saf.idx temp.${Population}/
cp ../WinSFS_Results/${Population}.saf.gz temp.${Population}/
cp ../WinSFS_Results/${Population}.saf.pos.gz temp.${Population}/

cd temp.${Population}/

head -2 ${Population}.winsfs_folded.sfs | tail -1 > ${Population}.winsfs_folded.converted.sfs

# Run command:
module purge 
module load angsd/0.935-GCC-10.2.0
Threads="16"

realSFS saf2theta ${Population}.saf.idx -outname ${Population} -sfs ${Population}.winsfs_folded.converted.sfs -fold 1 -P ${Threads}


#calculate Tajimas D globally/population wise per chromosome
thetaStat do_stat ${Population}.thetas.idx

rm ${Population}.winsfs_folded.sfs
rm ${Population}.saf.idx
rm ${Population}.saf.gz
rm ${Population}.saf.pos.gz

cp * ../

cd ../
rm -r temp.${Population}

```

