# Individual Heterozygosity + Inbreeding


```bash
mkdir Heterozygosity
cd Heterozygosity

cp ../bams_good .
cp ../Puffin_ReferenceGenome.NU.MT.FINAL.fasta .
cp ../sites2do .

mkdir SingleSFS

cat bams_good | awk -F '.' '{print $1}' | while read sample ;
do echo ${sample} ;
sbatch Hetero.sh ${sample} ;
done

```

nano Hetero.sh

```bash

mkdir temp.${1}_${SLURM_ARRAY_TASK_ID}
cd temp.${1}_${SLURM_ARRAY_TASK_ID}
cp ../bams_good .
cp ../Puffin_ReferenceGenome.NU.MT.FINAL.fasta .
cp ../ChromosomeList.edit .

Bamfile=$(cat bams_good | grep ${1})

# Run command:

module purge 
module load SAMtools/1.9-GCC-8.2.0-2.31.1
samtools faidx Puffin_ReferenceGenome.NU.MT.FINAL.fasta

module purge 
module load angsd/0.935-GCC-10.2.0

Threads="1"
GENOME_REF=Puffin_ReferenceGenome.NU.MT.FINAL.fasta

if [[ ${SLURM_ARRAY_TASK_ID} -gt 9 ]]
then
	ChromoNo=$(echo ${SLURM_ARRAY_TASK_ID} )
else
	ChromoNo=$(echo "0"${SLURM_ARRAY_TASK_ID} )
fi

Chromosome=$(cat ChromosomeList.edit | awk -F ':' '{print $1}' | grep -w Chr_${ChromoNo} )
ChromosomeEdit=$(echo ${Chromosome}":")
grep -w Chr_${ChromoNo} ../sites2do > sites_Chromosome${ChromoNo}_2do

sleep 120

angsd sites index sites_Chromosome${ChromoNo}_2do

sleep 120 

#run angsd
angsd -i ../../../${Bamfile} -anc ${GENOME_REF} -ref ${GENOME_REF} -doSaf 1 -GL 1 -r ${ChromosomeEdit} -sites sites_Chromosome${ChromoNo}_2do -doCounts 1 -uniqueOnly 1 -remove_bads 1 -C 50 -baq 2 -minMapQ 30 -minQ 30 -P ${Threads} -out ${1}_Chromosome${ChromoNo}

mv ${1}* ../SingleSFS/

cd ..
rm -r temp.${1}_${SLURM_ARRAY_TASK_ID}/

```

```bash

rm slurm* 

cd SingleSFS/

cat ../bams_good | awk -F '.' '{print $1}' | while read sample ;
do echo ${sample} ;
sbatch SFS.creation.sh ${sample} ;
done

```

SFS.creation.sh

```bash

module purge 
module load angsd/0.935-GCC-10.2.0
Threads="16"

Ind=$(echo ${1})

realSFS cat ${Ind}_Chromosome*.saf.idx -outnames ${Ind} ;
realSFS ${Ind}.saf.idx -P ${Threads} -fold 1 > ${Ind}.sfs ;

rm ${Ind}_Chromosome*.saf.gz
rm ${Ind}_Chromosome*.saf.idx
rm ${Ind}_Chromosome*.saf.pos.gz
rm ${Ind}_Chromosome*.arg 

module purge

winsfs -v -t ${Threads} ${Ind}.saf.idx > ${Ind}.winsfs.sfs

```

```bash


rm slurm*

module purge
module load R/3.6.0-intel-2019a

touch Heterozygosity.txt
cat ../bams_good | awk -F '.' '{print $1}' | while read sample ;
do echo ${sample} ;
tail -1 ${sample}.winsfs.sfs > tmp
Hetero=$(Rscript Hetero.R tmp) ;
echo -e ${sample}"\t"${Hetero} >> Heterozygosity.txt ;
done ;
rm tmp

```

Hetero.R

```r
infile <- commandArgs(trailingOnly = TRUE)
a<-scan(infile)
hetero <- a[2]/sum(a)
cat(hetero)
```


Inbreeding according to Kersten

```bash
mkdir Inbreeding
cd Inbreeding
cp ../Heterozygosity/bams_good .

cat bams_good | awk -F '.' '{print $1}' | while read sample ;
do cp Heterozygosity/SingleSFS/${sample}.* . ;
done

rm *.sfs

###### make Regions file
######
module purge 
module load angsd/0.935-GCC-10.2.0
Threads="16"
sample=BJH001
realSFS ${sample}.saf.idx ${sample}.saf.idx -P ${Threads} -fold 1 > test.ml
realSFS fst index ${sample}.saf.idx ${sample}.saf.idx -sfs test.ml -whichFst 1 -fstout test -fold 1
realSFS fst stats2 test.fst.idx -win 100000 -step 50000 -fold 1 > slidingwindow #this will determine RoH size
rm test.fst.gz
touch Regions.txt
cat slidingwindow | tail -n+2 | awk '{print $2}' | uniq | while read chromo ;
do echo ${chromo} ;
echo -e ${chromo}":1-100000" >> Regions.txt ;
grep -w ${chromo} slidingwindow | while read line ;
do Region=$(echo ${line} | awk '{print $1}' | awk -F '(' '{print $4}' | sed 's/)//g' | sed 's/,/-/g') ;
echo -e ${chromo}":"${Region} >> Regions.txt ;
done ;
done ;
######
wc -l Regions.txt


cat bams_good | awk -F '.' '{print $1}' | while read sample ;
do echo ${sample} ;
sbatch realSFS_ROH.sh ${sample} ;
done ;

```

nano realSFS_ROH.sh

```bash

module purge 
module load angsd/0.935-GCC-10.2.0
Threads="1"

touch ${1}_${SLURM_ARRAY_TASK_ID}.sfs

NumberOfLines=$(echo "475") 
BottomLine=$(bc <<<"scale=0; ${SLURM_ARRAY_TASK_ID} * 475 ") 

cat Regions.txt | head -${BottomLine} | tail -${NumberOfLines} | while read line ;
do realSFS ${1}.saf.idx -r ${line} -P ${Threads} -fold 1 >> ${1}_${SLURM_ARRAY_TASK_ID}.sfs
done ;

```

THEN

```bash

cat bams_good | awk -F '.' '{print $1}' | while read sample ;
do echo ${sample} ;
touch ${sample}.sfs ;
for i in {1..46} ;
do cat ${sample}_${i}.sfs >> ${sample}.sfs ;
rm ${sample}_${i}.sfs ;
done ;
done ;

nano HeteroRoh.R
### HeteroRoh.R ###
infile <- commandArgs(trailingOnly = TRUE)
sample <- substr(infile, 1, 6)
df <- as.data.frame(read.table(file = infile, stringsAsFactors = FALSE))
df$V4 <- rowSums(df)
df$V5 <- df$V2 / df$V4
df$V6 <- sample
sample_table <- as.data.frame(cbind(df$V6, df$V5))
write.table(sample_table, file = paste(sample,"heterozygosity.txt", sep = "_"), col.names = F, row.names = F, quote = F, sep = "\t")
###############

module purge
module load R/3.6.0-intel-2019a
cat bams_good | awk -F '.' '{print $1}' | while read sample ;
do echo ${sample} ;
paste Regions.txt ${sample}.sfs > ${sample}.regionSFS ;
Rscript HeteroRoh.R ${sample}.sfs ;
paste ${sample}.regionSFS ${sample}_heterozygosity.txt > ${sample}.RoH
done ;

cat *heterozygosity.txt > HeterozygosityCombined.txt

#choose cutoff
# plot data to assess cutoff


#then
nano ROH_Length_Count.R #change to RoH sizes
##### ROH_Length_Count.R
library(tidyr)
library(tidyverse)
infile <- commandArgs(trailingOnly = TRUE)
sample <- substr(infile, 1, 6)
df <- read_delim(file = infile, na = "", col_names = FALSE, delim="\t")
df_tmp <- as.data.frame(na.omit(df))
df_edit <- cbind(df_tmp$X3, data.frame(do.call('rbind', strsplit(as.character(df_tmp$X1),':',fixed=TRUE))), df_tmp$X4)
colnames(df_edit) <- c("Sample", "Chromosome", "Region", "Hetero")
df_edit$Chromosome <- sub("Chr_0", "Chr_", df_edit$Chromosome)
for (i in 1:24) {
df_edit2 <- df_edit %>% filter(Chromosome == paste("Chr_",i, sep = ""))
df_edit2$ROH <- ifelse(df_edit2$Hetero < 0.0004820091, 1, 0) # enter cutoff here
value <- rle(df_edit2$ROH)$values
lengths <- as.numeric(rle(df_edit2$ROH)$lengths)
df2 <- cbind.data.frame(value, lengths)
df2$chromosome <- paste("Chr",i, sep = "_")
df3 <- df2 %>% filter(value == 1) %>% filter(lengths > 1)
df3$ROH_Length <- df3$lengths * 0.1 - ((df3$lengths - 1) * 0.05 )
assign(paste("Chromo",i,"df3", sep = "_"), df3)
}
Dataframes <- lapply(paste("Chromo_",1:24,"_df3", sep = ""), get)
All_chromoMerge_df3 <- bind_rows(Dataframes)
All_chromoMerge_df3$value <- sample
colnames(All_chromoMerge_df3) <- c("Sample", "Number", "Chromosome","ROH_Length")
write.table(All_chromoMerge_df3, file = paste(sample,"_ROH_Lengths.txt", sep = ""), row.names = F, quote = F, sep = "\t")
sample_inbreeding <- unname(colSums(All_chromoMerge_df3[,"ROH_Length", drop = FALSE]) / (nrow(df_edit) * 0.1 - ((nrow(df_edit) - 1) * 0.05)))
Inbreeding_table <- as.data.frame(cbind(sample, sample_inbreeding))
write.table(Inbreeding_table, file = paste(sample,"_F_ROH.txt", sep = ""), row.names = F, quote = F, sep = "\t")
###############

module purge
module load R/3.6.0-intel-2019a

echo -e "Sample\tNumber\tChromosome\tROH_Length" > AllSamples_ROH_Length
echo -e "Sample\tF_ROH" > AllSamples_F_ROH
cat bams_good | awk -F '.' '{print $1}' | while read sample ;
do echo ${sample} ;
Rscript ROH_Length_Count.R ${sample}.RoH ;
cat ${sample}_ROH_Lengths.txt | tail -n+2 >> AllSamples_ROH_Length ;
cat ${sample}_F_ROH.txt | tail -n+2 >> AllSamples_F_ROH ;
done ;

```